{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Binary Simplification in Metabolomics\n",
    "\n",
    "Notebook to support the study on the application of **Bin**ary **Sim**plification as a competing form of pre-processing procedure for high-resolution metabolomics data.\n",
    "\n",
    "This is notebook `paper_binsim.ipynb`\n",
    "\n",
    "\n",
    "## Organization of the Notebook\n",
    "\n",
    "- Set up database of data sets\n",
    "- Application of different pre-treatments (including BinSim) to each data set\n",
    "- Permutation tests generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Needed Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas.testing import assert_frame_equal\n",
    "\n",
    "import scipy.spatial.distance as dist\n",
    "import scipy.cluster.hierarchy as hier\n",
    "import scipy.stats as stats\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.patches as mpatches\n",
    "from matplotlib import ticker\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "# Metabolinks package\n",
    "import metabolinks as mtl\n",
    "import metabolinks.transformations as transf\n",
    "\n",
    "# Python files in the repository\n",
    "import multianalysis as ma\n",
    "from elips import plot_confidence_ellipse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Description of dataset records\n",
    "\n",
    "`datasets` is the global dict that holds all data sets. It is a **dict of dict's**\n",
    "\n",
    "Each data set is **represented as a dict**.\n",
    "\n",
    "TO CONSIDER: migrate the table of datasets to a pandas DataFrame.\n",
    "\n",
    "Each record has the following fields (keys):\n",
    "\n",
    "- `name`: the table/figure name of the data set\n",
    "- `source`: the biological source for each dataset\n",
    "- `mode`: the aquisition mode\n",
    "- `alignment`: the alignment used to generate the data matrix\n",
    "- `data`: the data matrix\n",
    "- `target`: the sample labels, possibly already integer encoded\n",
    "- `<treatment name>`: transformed data matrix. These treatment names can be\n",
    "    - `original`: an alias to `data`\n",
    "    - `imputed`: missing value imputed data, only\n",
    "    - `P`: Pareto scaled data\n",
    "    - `NP`: Pareto scaled and normalized\n",
    "    - `NGP`: normalized, glog transformed and Pareto scaled\n",
    "    - `BinSim`: binary simplified data\n",
    "\n",
    "The keys of `datasets` may be shared with dicts holding records resulting from comparison analysis.\n",
    "\n",
    "Here are the keys used in this study:\n",
    "\n",
    "- GD_neg_global2\n",
    "- GD_pos_global2\n",
    "- GD_neg_class2\n",
    "- GD_pos_class2\n",
    "- YD\n",
    "- YD2\n",
    "- vitis_types\n",
    "\n",
    "\n",
    "#### correspondence to previously used names/variables in more focused notebooks\n",
    "\n",
    "`aligned_all_pos` ↔ `datasets[GD_pos_global2]`\n",
    "\n",
    "`aligned_all_neg` ↔ `datasets[GD_neg_global2]`\n",
    "\n",
    "`aligned_all_pos2` ↔ `datasets[GD_pos_class2]`\n",
    "\n",
    "`aligned_all_neg2` ↔ `datasets[GD_neg_class2]`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Description of grapevine data sets\n",
    "\n",
    "Grapevine Datasets (Negative and Positive) - 33 samples belonging to 11 different grapevine varieties (3 samples per variety/biological group).\n",
    "\n",
    "5 different _Vitis_ species (other than _V. vinifera_) varieties:\n",
    "\n",
    "- CAN - 3 Samples (14, 15, 16) of _V. candicans Engelmann_ (VIVC variety number: 13508)\n",
    "- RIP - 3 Samples (17, 18, 19) of _V. riparia Michaux_ (Riparia Gloire de Montpellier, VIVC variety number: 4824) \n",
    "- ROT - 3 Samples (20, 21, 22) of _V. rotundifolia_ (Muscadinia Rotundifolia Michaux cv. Rotundifolia, VIVC variety number: 13586)\n",
    "- RU - 3 Samples (35, 36, 37) of _V. rupestris Scheele_ (Rupestris du lot, VIVC variety number: 10389)\n",
    "- LAB - 3 Samples (8, 9, 10) of _V. labrusca_ (Isabella, VIVC variety number: 5560)\n",
    "\n",
    "6 different _V. vinifera_ cultivars varieties are:\n",
    "\n",
    "- SYL - 3 samples (11, 12, 13) of the subspecies _sylvestris_ (VIVC variety number: -)\n",
    "- CS - 3 Samples (29, 30, 31) of the subspecies _sativa_ cultivar Cabernet Sauvignon (VIVC variety number: 1929)\n",
    "- PN - 3 Samples (23, 24, 25) of the subspecies _sativa_ cultivar Pinot Noir (VIVC variety number: 9279)\n",
    "- REG - 3 Samples (38, 39, 40) of the subspecies _sativa_ cultivar Regent (VIVC variety number: 4572)\n",
    "- RL - 3 Samples (26, 27, 28) of the subspecies _sativa_ cultivar Riesling Weiss (VIVC variety number: 10077)\n",
    "- TRI - 3 Samples (32, 33, 34) of the subspecies _sativa_ cultivar Cabernet Sauvignon (VIVC variety number: 15685)\n",
    "\n",
    "Data acquired by Marisa et al. (2020):\n",
    "\n",
    "Maia M, Ferreira AEN, Nascimento R, et al. Integrating metabolomics and targeted gene expression to uncover potential biomarkers of fungal / oomycetes ‑ associated disease susceptibility in grapevine. Sci Rep. Published online 2020:1-15. doi:10.1038/s41598-020-72781-2\n",
    "\n",
    "**Peak Alignment** and **Peak Filtering** was performed with function `metabolinks.peak_alignment.align()`\n",
    "(see `peak_alignments.ipynb` notebook)\n",
    "\n",
    "**4** data matrices were constructed from this data:\n",
    "\n",
    "- 1st Filtering/Alignment: Feature present in at least two samples in the dataset - min_sample = 2 considering all samples of the dataset. This alignment, \"1-2\", generated two datasets one for **positive mode** data acquisition and one for **negative mode** data acquisition. These data sets are names `GD_pos_global2` and `GD_neg_global2`, respectively.\n",
    "- 2nd Filtering/Alignment: Feature present in at least two of the three samples of any biological group in the dataset. \n",
    "Each group is defined by the biological replicates of each Vitis species/cultivar. This alignment, \"2-1\", generated two datasets one for **positive mode** data acquisition and one for **negative mode** data acquisition. These data sets are names `GD_pos_class2` and `GD_neg_class2`, respectively.\n",
    "\n",
    "For the purpose of assessing the performance of supervised methods each of these four datasets was used with target labels defining classes corresponding to replicates of each of the 11 Vitis species/cultivars.\n",
    "\n",
    "For the purpose of assessing the performance of supervised methods under a binary (two-class) problem, data set `GD_neg_class2` was also used with target labels defining two classes: Vitis vinifera cultivars and \"wild\", non-vinifera Vitis species. This is dataset `vitis_types`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Description of the yeast data set\n",
    "\n",
    "The Yeast dataset, data set `YD`, was acquired by J. Luz. The Yeast dataset consists of 15 samples of FT-ICR-MS metabolomics data obtained in positive ionization mode - 3 biological replicates of 5 different strains of Saccharomyces cerevisiae : the reference strain BY4741 (represented as BY) and 4 single-gene deletion mutants of this strain – ΔGLO1, ΔGLO2, ΔGRE3 and ΔENO1. These deleted genes are directly or indirectly related to methylglyoxal metabolism.\n",
    "\n",
    "Peak Alignment and Peak Filtering was performed with MetaboScape 4.0 software (see dissertation for details in sample preparation, pre-processing, formula assignment). In short, the peak filtering made was made by keeping all features that appeared in at least 2 samples of the full dataset.\n",
    "\n",
    "Yeast Dataset was obtained with Electrospray Ionization in Positive Mode, Human leucine enkephalin (Sigma Aldrich) is the reference feature (internal standard) used ([M+H]+ = 556.276575 Da or [M-H]- = 554.262022 Da) and pre-processed by MetaboScape 4.0 (Bruker Daltonics) - see dissertation for details in pre-processing and formula assignment.\n",
    "\n",
    "For the purpose of assessing the performance of supervised methods, this data set was used with target labels defining classes corresponding to replicates of each of the 4 yeast strains."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loading yeast data set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_MetScape_file(filename,\n",
    "                          col_renamer=None,\n",
    "                          add_labels=None,\n",
    "                          remove_ref_feat=None,):\n",
    "    \n",
    "    \"\"\"Read in a MetaboScape bucket table from a CSV file.\"\"\"\n",
    "    \n",
    "    data = pd.read_csv(filename).set_index('Bucket label')\n",
    "    \n",
    "    # optionally rename sample_names\n",
    "    if col_renamer is not None:\n",
    "        data = data.rename(columns=renamer)\n",
    "    \n",
    "    # optionally remove a reference feature (if already normalized)\n",
    "    if remove_ref_feat is not None:\n",
    "        #print(f'Feature {remove_ref_feat}\\n{data.loc[remove_ref_feat, :]}\\n----------')\n",
    "        data = data.drop(index=[remove_ref_feat])\n",
    "    \n",
    "    # split in peak metadata and intensities\n",
    "    peak_cols = ['m/z', 'Name', 'Formula']\n",
    "    intensity_cols = [c for c in list(data.columns) if c not in peak_cols]\n",
    "    peaks = data[peak_cols]\n",
    "    intensities = data[intensity_cols]\n",
    "\n",
    "    # replace zeros for NaN's\n",
    "    intensities = intensities.replace(0, np.nan).dropna(how='all')\n",
    "    \n",
    "    # force peaks to have the same features as the (trimmed) intensities\n",
    "    peaks = peaks.reindex(intensities.index)\n",
    "\n",
    "    # optionally, add labels to intensities\n",
    "    if add_labels is not None:\n",
    "        intensities = mtl.add_labels(intensities, labels=add_labels)\n",
    "    \n",
    "    return {'filename': filename,\n",
    "            'peaks':peaks,\n",
    "            'intensities': intensities}\n",
    "\n",
    "def renamer(colname):\n",
    "    # Util to optionally remove all those 00000 from sample names\n",
    "    return ''.join(colname.split('00000'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Labels of the 5 biological groups (5 yeast strains) - only added after\n",
    "yeast_classes = 'WT ΔGRE3 ΔENO1 ΔGLO1 ΔGLO2'.split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the file and keep results in dicts \n",
    "\n",
    "prefix_to_drop = None # change to 'ENO' to remove ENO strain\n",
    "\n",
    "# MScape non-normalized dataset\n",
    "MS_data = read_MetScape_file('5yeasts_notnorm.csv', \n",
    "                             remove_ref_feat=None,\n",
    "                             add_labels=None,\n",
    "                             col_renamer=renamer)\n",
    "\n",
    "#MS_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Bucket label</th>\n",
       "      <th>307.0838178877 Da</th>\n",
       "      <th>555.2692975341 Da</th>\n",
       "      <th>624.0873082632 Da</th>\n",
       "      <th>493.3168156897 Da</th>\n",
       "      <th>257.1028750516 Da</th>\n",
       "      <th>780.1092067055 Da</th>\n",
       "      <th>347.0630811124 Da</th>\n",
       "      <th>254.2246098628 Da</th>\n",
       "      <th>625.0869294704 Da</th>\n",
       "      <th>663.1091419334 Da</th>\n",
       "      <th>...</th>\n",
       "      <th>554.6687988435 Da</th>\n",
       "      <th>851.6819663479 Da</th>\n",
       "      <th>352.0866584613 Da</th>\n",
       "      <th>554.6387493588 Da</th>\n",
       "      <th>533.0467313042 Da</th>\n",
       "      <th>892.5313342202 Da</th>\n",
       "      <th>847.0028787316 Da</th>\n",
       "      <th>535.6351204857 Da</th>\n",
       "      <th>547.4252442854 Da</th>\n",
       "      <th>299.0852986149 Da</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BY0_1</th>\n",
       "      <td>391042880.0</td>\n",
       "      <td>111270160.0</td>\n",
       "      <td>130088936.0</td>\n",
       "      <td>106187168.0</td>\n",
       "      <td>54234476.0</td>\n",
       "      <td>33863468.0</td>\n",
       "      <td>35209136.0</td>\n",
       "      <td>9935943.0</td>\n",
       "      <td>25313166.0</td>\n",
       "      <td>21631346.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BY0_2</th>\n",
       "      <td>398003680.0</td>\n",
       "      <td>110423984.0</td>\n",
       "      <td>127133608.0</td>\n",
       "      <td>109233816.0</td>\n",
       "      <td>53720908.0</td>\n",
       "      <td>32622178.0</td>\n",
       "      <td>35636648.0</td>\n",
       "      <td>17679414.0</td>\n",
       "      <td>24989162.0</td>\n",
       "      <td>24420090.0</td>\n",
       "      <td>...</td>\n",
       "      <td>557228.3750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BY0_3</th>\n",
       "      <td>399481952.0</td>\n",
       "      <td>111225424.0</td>\n",
       "      <td>131208568.0</td>\n",
       "      <td>111553216.0</td>\n",
       "      <td>55826148.0</td>\n",
       "      <td>34655496.0</td>\n",
       "      <td>37943456.0</td>\n",
       "      <td>19530026.0</td>\n",
       "      <td>24684334.0</td>\n",
       "      <td>24862190.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRE3_1</th>\n",
       "      <td>270906400.0</td>\n",
       "      <td>92849040.0</td>\n",
       "      <td>130682664.0</td>\n",
       "      <td>104352720.0</td>\n",
       "      <td>46503348.0</td>\n",
       "      <td>33206608.0</td>\n",
       "      <td>32640452.0</td>\n",
       "      <td>5713388.5</td>\n",
       "      <td>25890872.0</td>\n",
       "      <td>28692028.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>548196.50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRE3_2</th>\n",
       "      <td>271023520.0</td>\n",
       "      <td>94888320.0</td>\n",
       "      <td>130844488.0</td>\n",
       "      <td>107277600.0</td>\n",
       "      <td>44828256.0</td>\n",
       "      <td>34597444.0</td>\n",
       "      <td>31696134.0</td>\n",
       "      <td>10139615.0</td>\n",
       "      <td>25548978.0</td>\n",
       "      <td>26964150.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRE3_3</th>\n",
       "      <td>274854272.0</td>\n",
       "      <td>94514272.0</td>\n",
       "      <td>130896568.0</td>\n",
       "      <td>109619152.0</td>\n",
       "      <td>46361588.0</td>\n",
       "      <td>33131336.0</td>\n",
       "      <td>31668180.0</td>\n",
       "      <td>16434200.0</td>\n",
       "      <td>25518902.0</td>\n",
       "      <td>26654308.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENO1_1</th>\n",
       "      <td>255276736.0</td>\n",
       "      <td>150620544.0</td>\n",
       "      <td>136612464.0</td>\n",
       "      <td>31276910.0</td>\n",
       "      <td>72465536.0</td>\n",
       "      <td>35298204.0</td>\n",
       "      <td>31236114.0</td>\n",
       "      <td>10759292.0</td>\n",
       "      <td>25798230.0</td>\n",
       "      <td>22221992.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>540980.9375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>543241.1875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>551425.1250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENO1_2</th>\n",
       "      <td>251610656.0</td>\n",
       "      <td>148516304.0</td>\n",
       "      <td>134972800.0</td>\n",
       "      <td>31337604.0</td>\n",
       "      <td>70376640.0</td>\n",
       "      <td>35198252.0</td>\n",
       "      <td>30789652.0</td>\n",
       "      <td>25896608.0</td>\n",
       "      <td>26849502.0</td>\n",
       "      <td>23433602.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>562062.625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>549861.4375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>553295.750</td>\n",
       "      <td>545085.3750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENO1_3</th>\n",
       "      <td>251024624.0</td>\n",
       "      <td>150645840.0</td>\n",
       "      <td>135922432.0</td>\n",
       "      <td>31644646.0</td>\n",
       "      <td>68912816.0</td>\n",
       "      <td>34380756.0</td>\n",
       "      <td>30916958.0</td>\n",
       "      <td>29723808.0</td>\n",
       "      <td>26606410.0</td>\n",
       "      <td>23500090.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>540392.625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>550756.625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>545906.1875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dGLO1_1</th>\n",
       "      <td>283822048.0</td>\n",
       "      <td>202869216.0</td>\n",
       "      <td>130259064.0</td>\n",
       "      <td>39904768.0</td>\n",
       "      <td>58269456.0</td>\n",
       "      <td>32095004.0</td>\n",
       "      <td>29669478.0</td>\n",
       "      <td>26538900.0</td>\n",
       "      <td>26013280.0</td>\n",
       "      <td>21245436.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>558994.1250</td>\n",
       "      <td>548675.125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dGLO1_2</th>\n",
       "      <td>285664416.0</td>\n",
       "      <td>205537504.0</td>\n",
       "      <td>128839264.0</td>\n",
       "      <td>40704944.0</td>\n",
       "      <td>56775068.0</td>\n",
       "      <td>32852638.0</td>\n",
       "      <td>30652686.0</td>\n",
       "      <td>27920220.0</td>\n",
       "      <td>24769882.0</td>\n",
       "      <td>21539278.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>563562.1875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dGLO1_3</th>\n",
       "      <td>295673952.0</td>\n",
       "      <td>209370176.0</td>\n",
       "      <td>132857032.0</td>\n",
       "      <td>40875840.0</td>\n",
       "      <td>58019620.0</td>\n",
       "      <td>35047176.0</td>\n",
       "      <td>30938974.0</td>\n",
       "      <td>29449846.0</td>\n",
       "      <td>26228898.0</td>\n",
       "      <td>21589446.0</td>\n",
       "      <td>...</td>\n",
       "      <td>547914.1875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>546448.6875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLO2_1</th>\n",
       "      <td>287722656.0</td>\n",
       "      <td>421969440.0</td>\n",
       "      <td>131498984.0</td>\n",
       "      <td>56428384.0</td>\n",
       "      <td>59549848.0</td>\n",
       "      <td>33618800.0</td>\n",
       "      <td>34360160.0</td>\n",
       "      <td>9995730.0</td>\n",
       "      <td>25419526.0</td>\n",
       "      <td>26483844.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>542414.5625</td>\n",
       "      <td>539765.375</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLO2_2</th>\n",
       "      <td>287749600.0</td>\n",
       "      <td>416559200.0</td>\n",
       "      <td>131380096.0</td>\n",
       "      <td>57348836.0</td>\n",
       "      <td>60346920.0</td>\n",
       "      <td>32398028.0</td>\n",
       "      <td>35621036.0</td>\n",
       "      <td>13618347.0</td>\n",
       "      <td>25224294.0</td>\n",
       "      <td>26875734.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>548258.75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLO2_3</th>\n",
       "      <td>290416512.0</td>\n",
       "      <td>419548864.0</td>\n",
       "      <td>131334040.0</td>\n",
       "      <td>58999288.0</td>\n",
       "      <td>61914296.0</td>\n",
       "      <td>33372098.0</td>\n",
       "      <td>36017192.0</td>\n",
       "      <td>19396972.0</td>\n",
       "      <td>25047738.0</td>\n",
       "      <td>25212104.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 1973 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Bucket label  307.0838178877 Da  555.2692975341 Da  624.0873082632 Da  \\\n",
       "BY0_1               391042880.0        111270160.0        130088936.0   \n",
       "BY0_2               398003680.0        110423984.0        127133608.0   \n",
       "BY0_3               399481952.0        111225424.0        131208568.0   \n",
       "GRE3_1              270906400.0         92849040.0        130682664.0   \n",
       "GRE3_2              271023520.0         94888320.0        130844488.0   \n",
       "GRE3_3              274854272.0         94514272.0        130896568.0   \n",
       "ENO1_1              255276736.0        150620544.0        136612464.0   \n",
       "ENO1_2              251610656.0        148516304.0        134972800.0   \n",
       "ENO1_3              251024624.0        150645840.0        135922432.0   \n",
       "dGLO1_1             283822048.0        202869216.0        130259064.0   \n",
       "dGLO1_2             285664416.0        205537504.0        128839264.0   \n",
       "dGLO1_3             295673952.0        209370176.0        132857032.0   \n",
       "GLO2_1              287722656.0        421969440.0        131498984.0   \n",
       "GLO2_2              287749600.0        416559200.0        131380096.0   \n",
       "GLO2_3              290416512.0        419548864.0        131334040.0   \n",
       "\n",
       "Bucket label  493.3168156897 Da  257.1028750516 Da  780.1092067055 Da  \\\n",
       "BY0_1               106187168.0         54234476.0         33863468.0   \n",
       "BY0_2               109233816.0         53720908.0         32622178.0   \n",
       "BY0_3               111553216.0         55826148.0         34655496.0   \n",
       "GRE3_1              104352720.0         46503348.0         33206608.0   \n",
       "GRE3_2              107277600.0         44828256.0         34597444.0   \n",
       "GRE3_3              109619152.0         46361588.0         33131336.0   \n",
       "ENO1_1               31276910.0         72465536.0         35298204.0   \n",
       "ENO1_2               31337604.0         70376640.0         35198252.0   \n",
       "ENO1_3               31644646.0         68912816.0         34380756.0   \n",
       "dGLO1_1              39904768.0         58269456.0         32095004.0   \n",
       "dGLO1_2              40704944.0         56775068.0         32852638.0   \n",
       "dGLO1_3              40875840.0         58019620.0         35047176.0   \n",
       "GLO2_1               56428384.0         59549848.0         33618800.0   \n",
       "GLO2_2               57348836.0         60346920.0         32398028.0   \n",
       "GLO2_3               58999288.0         61914296.0         33372098.0   \n",
       "\n",
       "Bucket label  347.0630811124 Da  254.2246098628 Da  625.0869294704 Da  \\\n",
       "BY0_1                35209136.0          9935943.0         25313166.0   \n",
       "BY0_2                35636648.0         17679414.0         24989162.0   \n",
       "BY0_3                37943456.0         19530026.0         24684334.0   \n",
       "GRE3_1               32640452.0          5713388.5         25890872.0   \n",
       "GRE3_2               31696134.0         10139615.0         25548978.0   \n",
       "GRE3_3               31668180.0         16434200.0         25518902.0   \n",
       "ENO1_1               31236114.0         10759292.0         25798230.0   \n",
       "ENO1_2               30789652.0         25896608.0         26849502.0   \n",
       "ENO1_3               30916958.0         29723808.0         26606410.0   \n",
       "dGLO1_1              29669478.0         26538900.0         26013280.0   \n",
       "dGLO1_2              30652686.0         27920220.0         24769882.0   \n",
       "dGLO1_3              30938974.0         29449846.0         26228898.0   \n",
       "GLO2_1               34360160.0          9995730.0         25419526.0   \n",
       "GLO2_2               35621036.0         13618347.0         25224294.0   \n",
       "GLO2_3               36017192.0         19396972.0         25047738.0   \n",
       "\n",
       "Bucket label  663.1091419334 Da  ...  554.6687988435 Da  851.6819663479 Da  \\\n",
       "BY0_1                21631346.0  ...                NaN                NaN   \n",
       "BY0_2                24420090.0  ...        557228.3750                NaN   \n",
       "BY0_3                24862190.0  ...                NaN                NaN   \n",
       "GRE3_1               28692028.0  ...                NaN                NaN   \n",
       "GRE3_2               26964150.0  ...                NaN                NaN   \n",
       "GRE3_3               26654308.0  ...                NaN                NaN   \n",
       "ENO1_1               22221992.0  ...                NaN        540980.9375   \n",
       "ENO1_2               23433602.0  ...                NaN                NaN   \n",
       "ENO1_3               23500090.0  ...                NaN                NaN   \n",
       "dGLO1_1              21245436.0  ...                NaN                NaN   \n",
       "dGLO1_2              21539278.0  ...                NaN        563562.1875   \n",
       "dGLO1_3              21589446.0  ...        547914.1875                NaN   \n",
       "GLO2_1               26483844.0  ...                NaN                NaN   \n",
       "GLO2_2               26875734.0  ...                NaN                NaN   \n",
       "GLO2_3               25212104.0  ...                NaN                NaN   \n",
       "\n",
       "Bucket label  352.0866584613 Da  554.6387493588 Da  533.0467313042 Da  \\\n",
       "BY0_1                       NaN                NaN                NaN   \n",
       "BY0_2                       NaN                NaN                NaN   \n",
       "BY0_3                       NaN                NaN                NaN   \n",
       "GRE3_1                      NaN                NaN                NaN   \n",
       "GRE3_2                      NaN                NaN                NaN   \n",
       "GRE3_3                      NaN                NaN                NaN   \n",
       "ENO1_1                      NaN        543241.1875                NaN   \n",
       "ENO1_2               562062.625                NaN                NaN   \n",
       "ENO1_3               540392.625                NaN         550756.625   \n",
       "dGLO1_1                     NaN        558994.1250         548675.125   \n",
       "dGLO1_2                     NaN                NaN                NaN   \n",
       "dGLO1_3                     NaN                NaN                NaN   \n",
       "GLO2_1                      NaN                NaN                NaN   \n",
       "GLO2_2                      NaN                NaN                NaN   \n",
       "GLO2_3                      NaN                NaN                NaN   \n",
       "\n",
       "Bucket label  892.5313342202 Da  847.0028787316 Da  535.6351204857 Da  \\\n",
       "BY0_1                       NaN                NaN                NaN   \n",
       "BY0_2                       NaN                NaN                NaN   \n",
       "BY0_3                       NaN                NaN                NaN   \n",
       "GRE3_1                548196.50                NaN                NaN   \n",
       "GRE3_2                      NaN                NaN                NaN   \n",
       "GRE3_3                      NaN                NaN                NaN   \n",
       "ENO1_1                      NaN                NaN        551425.1250   \n",
       "ENO1_2                      NaN        549861.4375                NaN   \n",
       "ENO1_3                      NaN                NaN                NaN   \n",
       "dGLO1_1                     NaN                NaN                NaN   \n",
       "dGLO1_2                     NaN                NaN                NaN   \n",
       "dGLO1_3                     NaN        546448.6875                NaN   \n",
       "GLO2_1                      NaN                NaN        542414.5625   \n",
       "GLO2_2                548258.75                NaN                NaN   \n",
       "GLO2_3                      NaN                NaN                NaN   \n",
       "\n",
       "Bucket label  547.4252442854 Da  299.0852986149 Da  \n",
       "BY0_1                       NaN                NaN  \n",
       "BY0_2                       NaN                NaN  \n",
       "BY0_3                       NaN                NaN  \n",
       "GRE3_1                      NaN                NaN  \n",
       "GRE3_2                      NaN                NaN  \n",
       "GRE3_3                      NaN                NaN  \n",
       "ENO1_1                      NaN                NaN  \n",
       "ENO1_2               553295.750        545085.3750  \n",
       "ENO1_3                      NaN        545906.1875  \n",
       "dGLO1_1                     NaN                NaN  \n",
       "dGLO1_2                     NaN                NaN  \n",
       "dGLO1_3                     NaN                NaN  \n",
       "GLO2_1               539765.375                NaN  \n",
       "GLO2_2                      NaN                NaN  \n",
       "GLO2_3                      NaN                NaN  \n",
       "\n",
       "[15 rows x 1973 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# keep features that appear in at least two samples\n",
    "yeast_datamatrix = transf.keep_atleast(MS_data['intensities'].transpose(), minimum=2)\n",
    "yeast_datamatrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Building data-sets data base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target for grapevine 11-variety data sets\n",
      "['CAN', 'CAN', 'CAN', 'CS', 'CS', 'CS', 'LAB', 'LAB', 'LAB', 'PN', 'PN', 'PN', 'REG', 'REG', 'REG', 'RIP', 'RIP', 'RIP', 'RL', 'RL', 'RL', 'ROT', 'ROT', 'ROT', 'RU', 'RU', 'RU', 'SYL', 'SYL', 'SYL', 'TRI', 'TRI', 'TRI']\n",
      "------\n",
      "target for 4 yeast strains data set\n",
      "['WT', 'WT', 'WT', 'ΔGRE3', 'ΔGRE3', 'ΔGRE3', 'ΔENO1', 'ΔENO1', 'ΔENO1', 'ΔGLO1', 'ΔGLO1', 'ΔGLO1', 'ΔGLO2', 'ΔGLO2', 'ΔGLO2']\n",
      "------\n",
      "target for 2-class wild Vitis vs Vitis vinifera data set\n",
      "['wild', 'wild', 'wild', 'vinifera', 'vinifera', 'vinifera', 'wild', 'wild', 'wild', 'vinifera', 'vinifera', 'vinifera', 'vinifera', 'vinifera', 'vinifera', 'wild', 'wild', 'wild', 'vinifera', 'vinifera', 'vinifera', 'wild', 'wild', 'wild', 'wild', 'wild', 'wild', 'wild', 'wild', 'wild', 'vinifera', 'vinifera', 'vinifera']\n"
     ]
    }
   ],
   "source": [
    "datasets = {}\n",
    "# From the alignments_new store\n",
    "\n",
    "data_df = pd.HDFStore('alignments_new.h5').get('all_1ppm_min2_neg').transpose()\n",
    "gd_labels = mtl.parse_data(data_df, labels_loc='label').sample_labels\n",
    "\n",
    "datasets['GD_neg_global2'] = {'source': 'grapevine',\n",
    "                              'alignment': '1-2',\n",
    "                              'mode': '-',\n",
    "                              'name': 'GDg2-',\n",
    "                              'data': data_df,\n",
    "                              'original': data_df,\n",
    "                              'target': gd_labels,\n",
    "                              'classes': list(pd.unique(gd_labels))}\n",
    "\n",
    "\n",
    "data_df = pd.HDFStore('alignments_new.h5').get('all_1ppm_min2_pos').transpose()\n",
    "labels = mtl.parse_data(data_df, labels_loc='label').sample_labels\n",
    "\n",
    "datasets['GD_pos_global2'] = {'source': 'grapevine',\n",
    "                              'alignment': '1-2',\n",
    "                              'mode': '+',\n",
    "                              'name': 'GDg2+',\n",
    "                              'data': data_df,\n",
    "                              'original': data_df,\n",
    "                              'target': gd_labels,\n",
    "                              'classes': list(pd.unique(gd_labels))}\n",
    "\n",
    "data_df = pd.HDFStore('alignments_new.h5').get('groups_1ppm_min2_all_1ppm_neg').transpose()\n",
    "labels = mtl.parse_data(data_df, labels_loc='label').sample_labels\n",
    "\n",
    "datasets['GD_neg_class2'] = {'source': 'grapevine',\n",
    "                              'alignment': '2-1',\n",
    "                              'mode': '-',\n",
    "                              'name': 'GDc2-',\n",
    "                              'data': data_df,\n",
    "                              'original': data_df,\n",
    "                              'target': gd_labels,\n",
    "                              'classes': list(pd.unique(gd_labels))}\n",
    "\n",
    "data_df = pd.HDFStore('alignments_new.h5').get('groups_1ppm_min2_all_1ppm_pos').transpose()\n",
    "labels = mtl.parse_data(data_df, labels_loc='label').sample_labels\n",
    "\n",
    "datasets['GD_pos_class2'] = {'source': 'grapevine',\n",
    "                              'alignment': '2-1',\n",
    "                              'mode': '+',\n",
    "                              'name': 'GDc2+',\n",
    "                              'data': data_df,\n",
    "                              'original': data_df,\n",
    "                              'target': gd_labels,\n",
    "                              'classes': list(pd.unique(gd_labels))}\n",
    "\n",
    "yeast_labels = [item for item in yeast_classes for i in range(3)]\n",
    "\n",
    "datasets['YD'] = {'source': 'yeast',\n",
    "                            'alignment': '1-2',\n",
    "                            'mode': '+',\n",
    "                            'name': 'YD 2/15',\n",
    "                            'data': yeast_datamatrix,\n",
    "                            'original': yeast_datamatrix,\n",
    "                            'target': yeast_labels,\n",
    "                            'classes': list(pd.unique(yeast_labels))}\n",
    "\n",
    "min_non_na = '6/15'\n",
    "\n",
    "yeast_datamatrix_less_na = transf.keep_atleast(yeast_datamatrix, minimum=eval(min_non_na))\n",
    "\n",
    "datasets['YD2'] = {'source': 'yeast',\n",
    "                            f'alignment': 'at_least {min_non_na}',\n",
    "                            'mode': '+',\n",
    "                            'name': 'YD ' + min_non_na,\n",
    "                            'data': yeast_datamatrix_less_na,\n",
    "                            'original': yeast_datamatrix_less_na,\n",
    "                            'target': yeast_labels,\n",
    "                            'classes': list(pd.unique(yeast_labels))}\n",
    "\n",
    "vitis_types = {'CAN': 'wild', 'RIP': 'wild', 'ROT': 'wild','RU': 'wild', 'LAB': 'wild',\n",
    "               'SYL': 'wild','REG': 'vinifera','CS': 'vinifera','PN': 'vinifera','RL': 'vinifera',\n",
    "               'TRI': 'vinifera'}\n",
    "\n",
    "gd_type_labels = [vitis_types[lbl] for lbl in gd_labels]\n",
    "\n",
    "datasets['vitis_types'] = {'source': 'grapevine',\n",
    "                            'alignment': '2-1',\n",
    "                            'mode': '-',\n",
    "                            'name': 'Vitis types',\n",
    "                            'data': datasets['GD_neg_class2']['original'],\n",
    "                            'original': datasets['GD_neg_class2']['original'],\n",
    "                            'target': gd_type_labels,\n",
    "                            'classes': list(pd.unique(gd_type_labels))}\n",
    "\n",
    "print('target for grapevine 11-variety data sets')\n",
    "print(datasets['GD_neg_global2']['target'])\n",
    "print('------\\ntarget for 4 yeast strains data set')\n",
    "print(datasets['YD']['target'])\n",
    "print('------\\ntarget for 2-class wild Vitis vs Vitis vinifera data set')\n",
    "print(datasets['vitis_types']['target'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Bucket label</th>\n",
       "      <th>307.0838178877 Da</th>\n",
       "      <th>555.2692975341 Da</th>\n",
       "      <th>624.0873082632 Da</th>\n",
       "      <th>493.3168156897 Da</th>\n",
       "      <th>257.1028750516 Da</th>\n",
       "      <th>780.1092067055 Da</th>\n",
       "      <th>347.0630811124 Da</th>\n",
       "      <th>254.2246098628 Da</th>\n",
       "      <th>625.0869294704 Da</th>\n",
       "      <th>663.1091419334 Da</th>\n",
       "      <th>...</th>\n",
       "      <th>229.2123516866 Da</th>\n",
       "      <th>394.1157975983 Da</th>\n",
       "      <th>379.1049581299 Da</th>\n",
       "      <th>248.0382813898 Da</th>\n",
       "      <th>355.6157973903 Da</th>\n",
       "      <th>234.1618657562 Da</th>\n",
       "      <th>384.2188536141 Da</th>\n",
       "      <th>561.1740363894 Da</th>\n",
       "      <th>384.1550071491 Da</th>\n",
       "      <th>486.1583934820 Da</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BY0_1</th>\n",
       "      <td>391042880.0</td>\n",
       "      <td>111270160.0</td>\n",
       "      <td>130088936.0</td>\n",
       "      <td>106187168.0</td>\n",
       "      <td>54234476.0</td>\n",
       "      <td>33863468.0</td>\n",
       "      <td>35209136.0</td>\n",
       "      <td>9935943.0</td>\n",
       "      <td>25313166.0</td>\n",
       "      <td>21631346.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>772182.5000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>594449.9375</td>\n",
       "      <td>698648.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>704317.1875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BY0_2</th>\n",
       "      <td>398003680.0</td>\n",
       "      <td>110423984.0</td>\n",
       "      <td>127133608.0</td>\n",
       "      <td>109233816.0</td>\n",
       "      <td>53720908.0</td>\n",
       "      <td>32622178.0</td>\n",
       "      <td>35636648.0</td>\n",
       "      <td>17679414.0</td>\n",
       "      <td>24989162.0</td>\n",
       "      <td>24420090.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>601594.2500</td>\n",
       "      <td>616405.1875</td>\n",
       "      <td>573408.3125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>591270.6875</td>\n",
       "      <td>594117.8125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BY0_3</th>\n",
       "      <td>399481952.0</td>\n",
       "      <td>111225424.0</td>\n",
       "      <td>131208568.0</td>\n",
       "      <td>111553216.0</td>\n",
       "      <td>55826148.0</td>\n",
       "      <td>34655496.0</td>\n",
       "      <td>37943456.0</td>\n",
       "      <td>19530026.0</td>\n",
       "      <td>24684334.0</td>\n",
       "      <td>24862190.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>748387.5625</td>\n",
       "      <td>661618.0625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>577509.4375</td>\n",
       "      <td>628977.9375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>592246.8125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRE3_1</th>\n",
       "      <td>270906400.0</td>\n",
       "      <td>92849040.0</td>\n",
       "      <td>130682664.0</td>\n",
       "      <td>104352720.0</td>\n",
       "      <td>46503348.0</td>\n",
       "      <td>33206608.0</td>\n",
       "      <td>32640452.0</td>\n",
       "      <td>5713388.5</td>\n",
       "      <td>25890872.0</td>\n",
       "      <td>28692028.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>670217.1250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>582062.5000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>561696.1875</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRE3_2</th>\n",
       "      <td>271023520.0</td>\n",
       "      <td>94888320.0</td>\n",
       "      <td>130844488.0</td>\n",
       "      <td>107277600.0</td>\n",
       "      <td>44828256.0</td>\n",
       "      <td>34597444.0</td>\n",
       "      <td>31696134.0</td>\n",
       "      <td>10139615.0</td>\n",
       "      <td>25548978.0</td>\n",
       "      <td>26964150.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>883340.5000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>601161.2500</td>\n",
       "      <td>559238.4375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRE3_3</th>\n",
       "      <td>274854272.0</td>\n",
       "      <td>94514272.0</td>\n",
       "      <td>130896568.0</td>\n",
       "      <td>109619152.0</td>\n",
       "      <td>46361588.0</td>\n",
       "      <td>33131336.0</td>\n",
       "      <td>31668180.0</td>\n",
       "      <td>16434200.0</td>\n",
       "      <td>25518902.0</td>\n",
       "      <td>26654308.0</td>\n",
       "      <td>...</td>\n",
       "      <td>562025.3125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>556031.8125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>681038.2500</td>\n",
       "      <td>558403.9375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENO1_1</th>\n",
       "      <td>255276736.0</td>\n",
       "      <td>150620544.0</td>\n",
       "      <td>136612464.0</td>\n",
       "      <td>31276910.0</td>\n",
       "      <td>72465536.0</td>\n",
       "      <td>35298204.0</td>\n",
       "      <td>31236114.0</td>\n",
       "      <td>10759292.0</td>\n",
       "      <td>25798230.0</td>\n",
       "      <td>22221992.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>580237.7500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENO1_2</th>\n",
       "      <td>251610656.0</td>\n",
       "      <td>148516304.0</td>\n",
       "      <td>134972800.0</td>\n",
       "      <td>31337604.0</td>\n",
       "      <td>70376640.0</td>\n",
       "      <td>35198252.0</td>\n",
       "      <td>30789652.0</td>\n",
       "      <td>25896608.0</td>\n",
       "      <td>26849502.0</td>\n",
       "      <td>23433602.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>657626.3125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENO1_3</th>\n",
       "      <td>251024624.0</td>\n",
       "      <td>150645840.0</td>\n",
       "      <td>135922432.0</td>\n",
       "      <td>31644646.0</td>\n",
       "      <td>68912816.0</td>\n",
       "      <td>34380756.0</td>\n",
       "      <td>30916958.0</td>\n",
       "      <td>29723808.0</td>\n",
       "      <td>26606410.0</td>\n",
       "      <td>23500090.0</td>\n",
       "      <td>...</td>\n",
       "      <td>597902.5000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>594469.9375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>664188.5000</td>\n",
       "      <td>597523.3125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dGLO1_1</th>\n",
       "      <td>283822048.0</td>\n",
       "      <td>202869216.0</td>\n",
       "      <td>130259064.0</td>\n",
       "      <td>39904768.0</td>\n",
       "      <td>58269456.0</td>\n",
       "      <td>32095004.0</td>\n",
       "      <td>29669478.0</td>\n",
       "      <td>26538900.0</td>\n",
       "      <td>26013280.0</td>\n",
       "      <td>21245436.0</td>\n",
       "      <td>...</td>\n",
       "      <td>664256.5000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>618080.8125</td>\n",
       "      <td>540651.7500</td>\n",
       "      <td>607250.8125</td>\n",
       "      <td>590962.6250</td>\n",
       "      <td>569196.3750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>704541.7500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dGLO1_2</th>\n",
       "      <td>285664416.0</td>\n",
       "      <td>205537504.0</td>\n",
       "      <td>128839264.0</td>\n",
       "      <td>40704944.0</td>\n",
       "      <td>56775068.0</td>\n",
       "      <td>32852638.0</td>\n",
       "      <td>30652686.0</td>\n",
       "      <td>27920220.0</td>\n",
       "      <td>24769882.0</td>\n",
       "      <td>21539278.0</td>\n",
       "      <td>...</td>\n",
       "      <td>834017.9375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>735004.6250</td>\n",
       "      <td>592491.3125</td>\n",
       "      <td>631597.5000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dGLO1_3</th>\n",
       "      <td>295673952.0</td>\n",
       "      <td>209370176.0</td>\n",
       "      <td>132857032.0</td>\n",
       "      <td>40875840.0</td>\n",
       "      <td>58019620.0</td>\n",
       "      <td>35047176.0</td>\n",
       "      <td>30938974.0</td>\n",
       "      <td>29449846.0</td>\n",
       "      <td>26228898.0</td>\n",
       "      <td>21589446.0</td>\n",
       "      <td>...</td>\n",
       "      <td>699665.1250</td>\n",
       "      <td>546246.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>550351.9375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>622481.3750</td>\n",
       "      <td>674023.5625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLO2_1</th>\n",
       "      <td>287722656.0</td>\n",
       "      <td>421969440.0</td>\n",
       "      <td>131498984.0</td>\n",
       "      <td>56428384.0</td>\n",
       "      <td>59549848.0</td>\n",
       "      <td>33618800.0</td>\n",
       "      <td>34360160.0</td>\n",
       "      <td>9995730.0</td>\n",
       "      <td>25419526.0</td>\n",
       "      <td>26483844.0</td>\n",
       "      <td>...</td>\n",
       "      <td>651178.9375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>758860.2500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>622542.1875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>719774.6875</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLO2_2</th>\n",
       "      <td>287749600.0</td>\n",
       "      <td>416559200.0</td>\n",
       "      <td>131380096.0</td>\n",
       "      <td>57348836.0</td>\n",
       "      <td>60346920.0</td>\n",
       "      <td>32398028.0</td>\n",
       "      <td>35621036.0</td>\n",
       "      <td>13618347.0</td>\n",
       "      <td>25224294.0</td>\n",
       "      <td>26875734.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>637697.0625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>667883.9375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>546216.5000</td>\n",
       "      <td>593309.5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLO2_3</th>\n",
       "      <td>290416512.0</td>\n",
       "      <td>419548864.0</td>\n",
       "      <td>131334040.0</td>\n",
       "      <td>58999288.0</td>\n",
       "      <td>61914296.0</td>\n",
       "      <td>33372098.0</td>\n",
       "      <td>36017192.0</td>\n",
       "      <td>19396972.0</td>\n",
       "      <td>25047738.0</td>\n",
       "      <td>25212104.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>642698.4375</td>\n",
       "      <td>693618.6250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>631244.3750</td>\n",
       "      <td>643995.1250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>588056.3125</td>\n",
       "      <td>632353.9375</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 606 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Bucket label  307.0838178877 Da  555.2692975341 Da  624.0873082632 Da  \\\n",
       "BY0_1               391042880.0        111270160.0        130088936.0   \n",
       "BY0_2               398003680.0        110423984.0        127133608.0   \n",
       "BY0_3               399481952.0        111225424.0        131208568.0   \n",
       "GRE3_1              270906400.0         92849040.0        130682664.0   \n",
       "GRE3_2              271023520.0         94888320.0        130844488.0   \n",
       "GRE3_3              274854272.0         94514272.0        130896568.0   \n",
       "ENO1_1              255276736.0        150620544.0        136612464.0   \n",
       "ENO1_2              251610656.0        148516304.0        134972800.0   \n",
       "ENO1_3              251024624.0        150645840.0        135922432.0   \n",
       "dGLO1_1             283822048.0        202869216.0        130259064.0   \n",
       "dGLO1_2             285664416.0        205537504.0        128839264.0   \n",
       "dGLO1_3             295673952.0        209370176.0        132857032.0   \n",
       "GLO2_1              287722656.0        421969440.0        131498984.0   \n",
       "GLO2_2              287749600.0        416559200.0        131380096.0   \n",
       "GLO2_3              290416512.0        419548864.0        131334040.0   \n",
       "\n",
       "Bucket label  493.3168156897 Da  257.1028750516 Da  780.1092067055 Da  \\\n",
       "BY0_1               106187168.0         54234476.0         33863468.0   \n",
       "BY0_2               109233816.0         53720908.0         32622178.0   \n",
       "BY0_3               111553216.0         55826148.0         34655496.0   \n",
       "GRE3_1              104352720.0         46503348.0         33206608.0   \n",
       "GRE3_2              107277600.0         44828256.0         34597444.0   \n",
       "GRE3_3              109619152.0         46361588.0         33131336.0   \n",
       "ENO1_1               31276910.0         72465536.0         35298204.0   \n",
       "ENO1_2               31337604.0         70376640.0         35198252.0   \n",
       "ENO1_3               31644646.0         68912816.0         34380756.0   \n",
       "dGLO1_1              39904768.0         58269456.0         32095004.0   \n",
       "dGLO1_2              40704944.0         56775068.0         32852638.0   \n",
       "dGLO1_3              40875840.0         58019620.0         35047176.0   \n",
       "GLO2_1               56428384.0         59549848.0         33618800.0   \n",
       "GLO2_2               57348836.0         60346920.0         32398028.0   \n",
       "GLO2_3               58999288.0         61914296.0         33372098.0   \n",
       "\n",
       "Bucket label  347.0630811124 Da  254.2246098628 Da  625.0869294704 Da  \\\n",
       "BY0_1                35209136.0          9935943.0         25313166.0   \n",
       "BY0_2                35636648.0         17679414.0         24989162.0   \n",
       "BY0_3                37943456.0         19530026.0         24684334.0   \n",
       "GRE3_1               32640452.0          5713388.5         25890872.0   \n",
       "GRE3_2               31696134.0         10139615.0         25548978.0   \n",
       "GRE3_3               31668180.0         16434200.0         25518902.0   \n",
       "ENO1_1               31236114.0         10759292.0         25798230.0   \n",
       "ENO1_2               30789652.0         25896608.0         26849502.0   \n",
       "ENO1_3               30916958.0         29723808.0         26606410.0   \n",
       "dGLO1_1              29669478.0         26538900.0         26013280.0   \n",
       "dGLO1_2              30652686.0         27920220.0         24769882.0   \n",
       "dGLO1_3              30938974.0         29449846.0         26228898.0   \n",
       "GLO2_1               34360160.0          9995730.0         25419526.0   \n",
       "GLO2_2               35621036.0         13618347.0         25224294.0   \n",
       "GLO2_3               36017192.0         19396972.0         25047738.0   \n",
       "\n",
       "Bucket label  663.1091419334 Da  ...  229.2123516866 Da  394.1157975983 Da  \\\n",
       "BY0_1                21631346.0  ...                NaN        772182.5000   \n",
       "BY0_2                24420090.0  ...                NaN        601594.2500   \n",
       "BY0_3                24862190.0  ...                NaN        748387.5625   \n",
       "GRE3_1               28692028.0  ...                NaN        670217.1250   \n",
       "GRE3_2               26964150.0  ...                NaN                NaN   \n",
       "GRE3_3               26654308.0  ...        562025.3125                NaN   \n",
       "ENO1_1               22221992.0  ...                NaN                NaN   \n",
       "ENO1_2               23433602.0  ...                NaN                NaN   \n",
       "ENO1_3               23500090.0  ...        597902.5000                NaN   \n",
       "dGLO1_1              21245436.0  ...        664256.5000                NaN   \n",
       "dGLO1_2              21539278.0  ...        834017.9375                NaN   \n",
       "dGLO1_3              21589446.0  ...        699665.1250        546246.0000   \n",
       "GLO2_1               26483844.0  ...        651178.9375                NaN   \n",
       "GLO2_2               26875734.0  ...                NaN                NaN   \n",
       "GLO2_3               25212104.0  ...                NaN        642698.4375   \n",
       "\n",
       "Bucket label  379.1049581299 Da  248.0382813898 Da  355.6157973903 Da  \\\n",
       "BY0_1                       NaN        594449.9375        698648.0000   \n",
       "BY0_2               616405.1875        573408.3125                NaN   \n",
       "BY0_3               661618.0625                NaN                NaN   \n",
       "GRE3_1                      NaN        582062.5000                NaN   \n",
       "GRE3_2                      NaN        883340.5000                NaN   \n",
       "GRE3_3                      NaN        556031.8125                NaN   \n",
       "ENO1_1                      NaN                NaN                NaN   \n",
       "ENO1_2                      NaN                NaN                NaN   \n",
       "ENO1_3              594469.9375                NaN                NaN   \n",
       "dGLO1_1                     NaN        618080.8125        540651.7500   \n",
       "dGLO1_2                     NaN                NaN        735004.6250   \n",
       "dGLO1_3                     NaN                NaN        550351.9375   \n",
       "GLO2_1              758860.2500                NaN        622542.1875   \n",
       "GLO2_2              637697.0625                NaN                NaN   \n",
       "GLO2_3              693618.6250                NaN        631244.3750   \n",
       "\n",
       "Bucket label  234.1618657562 Da  384.2188536141 Da  561.1740363894 Da  \\\n",
       "BY0_1                       NaN                NaN        704317.1875   \n",
       "BY0_2                       NaN        591270.6875        594117.8125   \n",
       "BY0_3               577509.4375        628977.9375                NaN   \n",
       "GRE3_1                      NaN                NaN                NaN   \n",
       "GRE3_2                      NaN                NaN                NaN   \n",
       "GRE3_3                      NaN                NaN                NaN   \n",
       "ENO1_1                      NaN                NaN        580237.7500   \n",
       "ENO1_2              657626.3125                NaN                NaN   \n",
       "ENO1_3                      NaN        664188.5000        597523.3125   \n",
       "dGLO1_1             607250.8125        590962.6250        569196.3750   \n",
       "dGLO1_2             592491.3125        631597.5000                NaN   \n",
       "dGLO1_3                     NaN        622481.3750        674023.5625   \n",
       "GLO2_1                      NaN                NaN                NaN   \n",
       "GLO2_2              667883.9375                NaN                NaN   \n",
       "GLO2_3              643995.1250                NaN                NaN   \n",
       "\n",
       "Bucket label  384.1550071491 Da  486.1583934820 Da  \n",
       "BY0_1                       NaN                NaN  \n",
       "BY0_2                       NaN                NaN  \n",
       "BY0_3                       NaN        592246.8125  \n",
       "GRE3_1              561696.1875                NaN  \n",
       "GRE3_2              601161.2500        559238.4375  \n",
       "GRE3_3              681038.2500        558403.9375  \n",
       "ENO1_1                      NaN                NaN  \n",
       "ENO1_2                      NaN                NaN  \n",
       "ENO1_3                      NaN                NaN  \n",
       "dGLO1_1                     NaN        704541.7500  \n",
       "dGLO1_2                     NaN                NaN  \n",
       "dGLO1_3                     NaN                NaN  \n",
       "GLO2_1              719774.6875                NaN  \n",
       "GLO2_2              546216.5000        593309.5000  \n",
       "GLO2_3              588056.3125        632353.9375  \n",
       "\n",
       "[15 rows x 606 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasets['YD2']['data']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Colors for plots to ensure consistency"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 11- variety grapevine data sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# customize label colors for 11 grapevine varieties\n",
    "\n",
    "colours = sns.color_palette('Blues', 3)\n",
    "colours.extend(sns.color_palette('Greens', 3))\n",
    "#colours = sns.cubehelix_palette(n_colors=6, start=2, rot=0, dark=0.2, light=.9, reverse=True)\n",
    "colours.extend(sns.color_palette('flare', 5))\n",
    "\n",
    "ordered_vitis_labels = ('CAN','RIP','ROT','RU','LAB','SYL','REG','CS','PN','RL','TRI')\n",
    "\n",
    "vitis_label_colors = {lbl: c for lbl, c in zip(ordered_vitis_labels, colours)}\n",
    "\n",
    "tab20bcols = sns.color_palette('tab20b', 20)\n",
    "tab20ccols = sns.color_palette('tab20c', 20)\n",
    "tab20cols = sns.color_palette('tab20', 20)\n",
    "tab10cols = sns.color_palette('tab10', 10)\n",
    "dark2cols = sns.color_palette('Dark2', 8)\n",
    "\n",
    "vitis_label_colors['RU'] = tab20bcols[8]\n",
    "vitis_label_colors['CAN'] = tab20ccols[5]\n",
    "vitis_label_colors['REG'] = tab10cols[3]\n",
    "\n",
    "for name in datasets:\n",
    "    if name.startswith('GD'):\n",
    "        datasets[name]['label_colors'] = vitis_label_colors\n",
    "        datasets[name]['sample_colors'] = [vitis_label_colors[lbl] for lbl in datasets[name]['target']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAncAAABVCAYAAAArMcJ2AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAJ4UlEQVR4nO3de6xlZXnH8e8PbwhTlYtWsc2MRSSKHaCoUYN1iHiNRvGGqGlsxKbx0oRG06oQSRSDTYj1fonGlsRUxIgRxcRbSEZjFMmMwGiooqKDGhnUqeAwQ5nHP9Y6ZLs558w+zNl77f36/SQnOWettTPPc9Ze7/s767InVYUkSZLacMjQBUiSJGn9GO4kSZIaYriTJElqiOFOkiSpIYY7SZKkhtx7LRsfffh9atORh06rlsH97H4bhy5hqnb/fu/QJUzNQ7hp6BKm6o4jHjB0CVP121v/MHQJU3XcrjuGLmGqcvhhQ5cwNXfsa/scyO372+5vz/6hK5iuX9/+611V9eDx5WsKd5uOPJTvnHPSuhU1b96w6SNDlzBVV2z9ydAlTM3r7zxv6BKm6pcvPX3oEqbqsm9tG7qEqbr04zuHLmGqDn38KUOXMDU/37lh6BKm6vrb2j1hA/CDxtPdu697343LLW87skuSJP2ZMdxJkiQ1xHAnSZLUEMOdJElSQwx3kiRJDTHcSZIkNcRwJ0mS1BDDnSRJUkMMd5IkSQ0x3EmSJDXEcCdJktQQw50kSVJDDHeSJEkNMdxJkiQ1xHAnSZLUEMOdJElSQwx3kiRJDTHcSZIkNcRwJ0mS1BDDnSRJUkMMd5IkSQ0x3EmSJDXEcCdJktQQw50kSVJDDHeSJEkNMdxJkiQ1xHAnSZLUEMOdJElSQwx3kiRJDTHcSZIkNcRwJ0mS1BDDnSRJUkMMd5IkSQ0x3EmSJDXEcCdJktQQw50kSVJDDHeSJEkNMdxJkiQ1xHAnSZLUEMOdJElSQwx3kiRJDTHcSZIkNcRwJ0mS1BDDnSRJUkMMd5IkSQ0x3EmSJDXEcCdJktQQw50kSVJDDHeSJEkNMdxJkiQ1xHAnSZLUEMOdJElSQwx3kiRJDUlVTb5xcjNw4/TKuZujgV0z/PdmqeXewP4Wnf0trpZ7A/tbdPa3vjZW1YPHF64p3M1aku9W1eOGrmMaWu4N7G/R2d/iark3sL9FZ3+z4WVZSZKkhhjuJEmSGjLv4e6jQxcwRS33Bva36OxvcbXcG9jforO/GZjre+4kSZK0NvN+5k6SJElrYLiTJElqyCDhLslDk3wqyQ1Jvp/kiiSP6tedk+T2JA8c2X5LkkryvJFlX0iyZfbVr02SO5NsT3JdksuTPKhfvinJdf33W5LsTrItyQ+SvG3Qog9gpZ76dSck+XqS/03ywyTnpfOP/Wu2J9mX5Nr++wsHbOWAVtl/W5J8YWzb/0ry4kEKXYMkt66y7j1JbkpyyMiyVyW5uf897EjymSSHzabaeybJW/tar+nr/lKSd42s35jkx0kelOTKJIN/dMEkDjCe7Bk5xrYn+Yd+3YYkH+rH221Jrk7ymkEbmdBKc0WS9/a/g2uTXJXkEUPXulZj+/LSpWOqn+suGtnujUnOH6zQgzDJ/Ldokhw1coz9qh8vl36ueel35uEuSYDLgCur6tiqegzwFuAv+03OAq4Czhh76U7grTMrdP3sqaqTquqxwG+A162w3daqOhl4HPDKJKfMrMK1W7anJPcHPg9cWFWPAk4Engy8tqo+0b/mJOAXwGn9z/8+TAsTm3T/Lbw+0J0B/Bz4+7HVl/S/hxOAfcCZs65vUkmeBDwX+Luq2gycDpwNPD/Jo/vN3gOcV1W/G6bKe2y19+MNS8dY/3Vxv/xjwG+B4/ox5lnAkbMte+1WmSvOBI4BNlfV39K9Z383WKH33Oi+3Af8c798L/DCJEcPV9q6aW78rKpbRuayDwPvHvn5tnnpd4gzd6cBd1TVh5cWVNX2qtqa5FhgA3AuXcgb9T1gd5Knz67Udfct4OGrbVBVtwFXA8fOpKKDN9rTy4FvVtWXAarqD8DrgXkPcJM64P5bcKcB1wEf4u7HHwBJ7g0cThcW5tXDgF1VtRegqnZV1U3AvwIfTPJs4C+q6pNDFrkODvh+7MfUJwDnVtV+gKq6uaretdrr5sSycwVwG/DLkX52VtU8vx8nsRV4ZP/9/9M9cXnOcOVMRevj57hB+x0i3D2WLrws5yzgf+je6McnecjY+nfQBb+Fk+RewNPozmyttt1RwBOBHbOo62As09MJjO3bqroB2JDkATMub11Nuv8W3NLxdxnw3CT3GVl3ZpLtwE10Z30un315E/sy8Nf9rQEfTPJUgKq6gu6v6YuB1w5Z4MFa4f147Nhl2afQHZPfWwpCC2alueLTwPP6Hi9KcvKM61pX/R9MzwauHVn8AeAVGbk9aZH9mYyfd5mHfuftgYqXAZ/qB6LPAi8ZXVlVWwH6QWtR3L+fFG+hmxS/ssJ2T0myjW5iurCq5jncrdRTgJU+W2dRP3NnpV6b6jPJfYHnAJ+rqv8Dvg08Y2STS/rLDg+lm4TeNPMiJ1RVtwKnAP8E3AxckuRV/eoPAFdV1fUDlXewVhtPxi/Lbh1/cX8v4vYkv5hRveuuqnYCxwNvBvYDX0vytGGrukeW9uV3gZ8BH19a0R+DFwP/Mkxp62bS+a8Vc9PvEOFuB93A+yeSbAaOA76S5Kd0QW+5S0MXsFj33u3pJ8WNwH05wD13VXXK6GWIObVSTzvo7hm8S5K/AW6tqt/PtML1s1KvtwBHjG17JIv7H2I/C3ggcG1//J3KMsdfdR+MeTl3vydvrlTVnVV1ZVW9je7WgBf1q/b3X4tq0vFkyfeBE/v7KamqC/rXL8KZ9GXnCoCq2ltVX6qqNwHvBF4wy8LWyZ6RIP6Gqto3tv4/gVfT3QaxqNb6fl10c9PvEOHu68D9Rp/WSvJ4uhucz6+qTf3XMcDDk2wcfXF/P9cRdDfrL4yq2k33V9gbxy53LaxlevokcGqS0+GuByzeC/zHcFWuj2V6/SFwzNIN+v379ERg+2BFHpyzgLOXjj/gEcAzsvxTsacCN8yyuLVIcnyS40YWnQTcOFA5UzHpeFJVP6I7M/SO/lIRSQ6lO8s+75adK5I8Nckx/c+HAJtpbP8CVNVv6C5Bv3roWg5Wi/Pfauah35mHu/4v/zOAp/ePt+8Azge20N3rM+oyujN44y4A/mqKZU5FVW2jezBkuZ4W0mhPVbUHeD5wbpLr6S7fXQW8f8AS181Yr3uBVwKf6E/Df4YuHO0esMRJHZZk58jXW4BnAl9c2qB/sOcbwNLHD53ZX867BjgZePvMq57cBuC/0310xjXAY+jGmNV8ceT3cenUK1wHy4wn4/fcLV3SOxs4CvhRkquBrwL/NvuK12aVuWIzcHn/0RLX0D2A0MQYs4yLgBaeml3u/Xr82Dj0klVevnCGnu/978ckSZIaMm8PVEiSJOkgGO4kSZIaYriTJElqiOFOkiSpIYY7SZKkhhjuJEmSGmK4kyRJasgfAYsleDw905CyAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 792x72 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.palplot(vitis_label_colors.values())\n",
    "new_ticks = plt.xticks(range(len(ordered_vitis_labels)), ordered_vitis_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5 yeast strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# customize label colors for 5 yeast strains\n",
    "\n",
    "colours = sns.color_palette('Set1', 5)\n",
    "yeast_label_colors = {lbl: c for lbl, c in zip(yeast_classes, colours)}\n",
    "datasets['YD']['label_colors'] = yeast_label_colors\n",
    "datasets['YD']['sample_colors'] = [yeast_label_colors[lbl] for lbl in datasets['YD']['target']]\n",
    "datasets['YD2']['label_colors'] = yeast_label_colors\n",
    "datasets['YD2']['sample_colors'] = [yeast_label_colors[lbl] for lbl in datasets['YD2']['target']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAASgAAABVCAYAAAD6zHmyAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAG7ElEQVR4nO3bX4xcZRnH8e+vbkKz0Yi0VTCUbqNGTBXbWiCi1JBwhyFbXEIaxTTBqMREwaARvamJ0SiQRsUYjYkbvWoTNCGmiCmpiYEFobWWLrTG2sUICpRegAXRwuPF+w4eZ2e33e7OnGfL75NM9rzvOWfmeTizv54/iyICM7OMlrRdgJnZTBxQZpaWA8rM0nJAmVlaDigzS2toLhufs2RJrBya0y6LyuHlI22X0FdDw0+3XULfvPnosrZL6KsVw0faLqGv9vydoxGxont+TmmzcmiInSvetnBVJfOxT32/7RL66q3rvtd2CX1z1U83t11CX31m3fVtl9BX+jpP9Jr3JZ6ZpeWAMrO0HFBmlpYDyszSckCZWVoOKDNLywFlZmk5oMwsLQeUmaXlgDKztBxQZpaWA8rM0nJAmVlaDigzS8sBZWZpOaDMLC0HlJml5YAys7QcUGaWlgPKzNJyQJlZWg4oM0vLAWVmaTmgzCwtB5SZpeWAMrO0HFBmlpYDyszSckCZWVoOKDNLywFlZmk5oMwsLQeUmaXlgDKztBxQZpaWA8rM0nJAmVlaDigzS8sBZWZpOaDMLC0HlJml5YAys7QcUGaWlgPKzNJyQJlZWg4oM0vLAWVmaTmgzCwtB5SZpeWAMrO0HFBmlpYDyszSckCZWVqKiFPfWHoWeKJ/5UyzHDg6wM8bpDO5N3B/i92g+1sVESu6J+cUUIMm6ZGI2NB2Hf1wJvcG7m+xy9KfL/HMLC0HlJmllT2gftx2AX10JvcG7m+xS9Ff6ntQZvb6lv0MysxexxxQZpZWqwElaZukmxrjeyX9pDG+S9LzkvZJOibpSF3e1UrBXSRtkhSSLmzMvUvSryQdlrRH0m5JG+u6LZKerT0clHRzY7+tkp6s6zqvsyVd0hj/UdKmtvqTNCLppa4aP1nXTUm6q7HvmKTxxnhU0v7a96OSRhvrrpU0KelVSQN7tH2ax+/OHu/zYUm/r70dlPTpxrqNkvZKOiFpbDCdvfbZg+jvi5Ieq8f2PkmrFrSJiGjtBVwL7KjLS4A9wERj/QRwaV0eB8barLdH/TuAh4GtdbwU+BNwdWOb9wJb6vIW4M66vIzyh3Ar63grcEuPzxgGhuryecAznXEL/Y0AB2bYdoryR7xr6ngMGK/L7wf+DKyu49V1fFEdvwd4N/BbYMNiOH6N9ecCfwXW1/Hy+j2+qvHf7CLgZ4P+/g6ovyuA4bp8I7B9IXto+xLvfuCyurwGOAC8IOktks6ifHH/0FZxs5H0RuAjwA3A5jr9cUrA3t3ZLiIORMR49/4R8Rzll/S82T4nIl6MiBN1uBQYyFONGfo7mduBr/aYvwX4ZkQcAag/vwV8qY4fj4hD8y56DuZ7/Bo+RwnivXX7o8CXga/U8VRE7AdeXfAmZjHA/nZHxIt12weB8xeyj1YDKiKeAk5IuoASVBPAQ8AHgQ3A/oj4d4slzmYU2FW/fMclraeE7N5T2bn2vBTY35i+uXHptLux7aWSJoFHgc82AqufRpneH8A7ui7xLm/sswNYL+mdXe+1hvKvbtMjdb4to8zj+DVk7A3a6e8G4J45vv+s2j6Dgv+dRXUCaqIxfqDFuk5mM+UXkvpz2lmGpF9KOiDpF43p62rY/AX4bkT8q7FuW0Ssra8rOpMR8VBErAEuBm6VtHTBu5lupv4ON2pcGxG/a+zzCnAbcGvXe4npZ3695gbpdI/ftM3o3Ufbf78z0P4kfYJyUnHb6ZXbW4aAeoASRu+jXOI9SDmDuowSXulIWgZcAvy6Tm0HrgMmgc6ZBhGxiXJdf05j9+01bC4H7pB07ql+bkQ8Dhyn3Dfom1n60yns/nNgI3BBY26S8uVtWg88Nr9KT888j1+3Xr19gJZ6g8H3J+lK4GuUe1svz7P8/5MhoO4HPgoci4hXIuIYcDYlpCbaLGwWY8DOzsGo91T+QbkB+SFJVze2He71BhExQfll/sJsHyRptaShuryKcjN5ar4NnMRM/Z30/kJE/AfYBtzUmL6dcuY3AuVpIOVe1R0LWfQczPv4NfwA2CJpLbwWDt8GvrPQRc/BwPqTtA74ESWcnlnIJoB2n+LVO/9vAJ4HvtGYGwcOdW03TpKneJSnTccoQdF5vQD8ELgQ2Em5hJsAfgNcGT2ekgBvp3xx3kR5ivcksK/xGgGup/wrto9y/2C0xf7uAV7qqvHzdZ8pYHldPgt4ivoUr85dQ7mHdrD+vKaxbhPwN+Bl4Gng3sTH75+11s7rfMoZ48O1t0PAjY3Purhudxx4DphM/v2ca3+76jHrfB/uXshe/L+6mFlaGS7xzMx6ckCZWVoOKDNLywFlZmk5oMwsLQeUmaXlgDKztP4LKkDcIUADbWsAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 360x72 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.palplot(yeast_label_colors.values())\n",
    "new_ticks = plt.xticks(range(len(yeast_classes)), yeast_classes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2 classes of Vitis types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# customize label colors for 2 types of Vitis varieties\n",
    "\n",
    "colours = [vitis_label_colors['SYL'], vitis_label_colors['TRI']]\n",
    "vitis_type_classes = datasets['vitis_types']['classes']\n",
    "vitis_types_label_colors = {lbl: c for lbl, c in zip(vitis_type_classes, colours)}\n",
    "datasets['vitis_types']['label_colors'] = vitis_types_label_colors\n",
    "datasets['vitis_types']['sample_colors'] = [vitis_types_label_colors[lbl] for lbl in datasets['vitis_types']['target']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAIEAAABVCAYAAABuD4dpAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAADzklEQVR4nO3dT4iUdRzH8fennWJjNdcYzT+VUCdPIi7YJUGCwJNgexDqYKciyIgivBkIQl66CEaHkGhRUQoyRTMlXVAoS1lNLaW2SCX/lbKhovLtML/gaZ1xdtx9/kCfFwz7m3l+88zvGd4888xcVhGB/b89UPYCrHyOwByBOQLDERhQ62RyV89DUZvanddaStd7+eGyl5Cba7eucf32dTXb1lEEtandPP7GwolZVQUt/Xhu2UvIzcCZLS23+ePAHIE5AsMRGI7AcASGIzAcgeEIDEdgOALDERiOwHAEhiMwHIHhCAxHYDgCwxEYjsBwBIYjMByB4QgMR2A4AsMRGI7AcASGIzAcgeEIDEdgOALDERiOwHAEhiMwHIHhCAxHYDgCwxEYjsBwBIYjMByB4QgMR2A4AsMRGI7AAHXyD7MlXQR+zW85d6kDlwp8vSIVfWxzImJasw0dRVA0SYcjoq/sdeShSsfmjwNzBFb9CD4sewE5qsyxVfqawIpR9TOBFcARWDUjkLRTUm8aj7SYs1FSf6ELa0HSLEnbxjAve1wrJZ2UNJD7Atuo/DWBpJGImNTk8Y3AFxHR9s2vIkmngCUR8csY59ci4nYeaynlTCDpHUkr0/h9SfvS+DlJn0gallQf9RxJWi/phKQdwPQSlo6k9yS9lrn/rqS3JB1P91dI+lTSLkmnJa3LzB2WVJf0AfAU8LmkNyX1SPpI0reSjkhamtnXVknbgS8lTZK0V9L3ko79O2/cIqLwG/AMsDWNB4FvgAeB1cArwDBQT9tH0t9lwB6gC5gF/AX0l7D2+cD+zP0TwCLgeLq/AvgZmAJ00/iZ/Ym0LXtc2fFa4KU07gV+AnrSvn4HHk3basAjaVwHzpDO5uO5lXVN8B2wQNJk4CZwCOgDnqURRTOLgE0RcScizgH7ClnpKBFxBJiergPmAX8Cv42atjcirkbEDRqRzGmz2+eBVZKOAl/TiOfJtG1PRFxJYwFrJQ0BXwGzgcfGeUjUxruD+xERtyQNAy8DB4EhYDHwNHDyXk/Nf3Vjsg3oB2YAm5tsv5kZ36H9+yzghYj48T8PSguBvzMPvQhMAxZk3sPuzpZ+tzK/HRwA3k5/B4FXgaPR+kr1ALBcUpekmTSiKctmYDmNECbiwnQ38LokAUia32LeFOBCCmAx7c8wY1JmBIPATOBQRPwB3KD1RwHAZ8Bp4BiwAdif+wpbiIgfgMnA2Yg4PwG7XEPjmmgoXWCuaTFvAOiTdJjGWeHUBLx29b8iWv4q+WORFcsRmCMwR2A4AsMRGI7AgH8ArSnY43GTS6IAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 144x72 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.palplot(datasets['vitis_types']['label_colors'].values())\n",
    "new_ticks = plt.xticks(range(len(datasets['vitis_types']['classes'])), datasets['vitis_types']['classes'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def styled_sample_labels(sample_names, sample_labels, label_colors):\n",
    "\n",
    "    meta_table = pd.DataFrame({'label': sample_labels,\n",
    "                               'sample': sample_names}).set_index('sample').T\n",
    "\n",
    "    def apply_label_color(val):\n",
    "        red, green, blue = label_colors[val]\n",
    "        red, green, blue = int(red*255), int(green*255), int(blue*255)   \n",
    "        hexcode = '#%02x%02x%02x' % (red, green, blue)\n",
    "        css = f'background-color: {hexcode}'\n",
    "        return css\n",
    "    \n",
    "    return meta_table.style.applymap(apply_label_color)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "#T_cac7f_row0_col0,#T_cac7f_row0_col1,#T_cac7f_row0_col2{\n",
       "            background-color:  #fd8d3c;\n",
       "        }#T_cac7f_row0_col3,#T_cac7f_row0_col4,#T_cac7f_row0_col5{\n",
       "            background-color:  #de5c5b;\n",
       "        }#T_cac7f_row0_col6,#T_cac7f_row0_col7,#T_cac7f_row0_col8{\n",
       "            background-color:  #73c375;\n",
       "        }#T_cac7f_row0_col9,#T_cac7f_row0_col10,#T_cac7f_row0_col11{\n",
       "            background-color:  #c14067;\n",
       "        }#T_cac7f_row0_col12,#T_cac7f_row0_col13,#T_cac7f_row0_col14{\n",
       "            background-color:  #d62728;\n",
       "        }#T_cac7f_row0_col15,#T_cac7f_row0_col16,#T_cac7f_row0_col17{\n",
       "            background-color:  #6aadd5;\n",
       "        }#T_cac7f_row0_col18,#T_cac7f_row0_col19,#T_cac7f_row0_col20{\n",
       "            background-color:  #9a356f;\n",
       "        }#T_cac7f_row0_col21,#T_cac7f_row0_col22,#T_cac7f_row0_col23{\n",
       "            background-color:  #2070b4;\n",
       "        }#T_cac7f_row0_col24,#T_cac7f_row0_col25,#T_cac7f_row0_col26{\n",
       "            background-color:  #8c6d31;\n",
       "        }#T_cac7f_row0_col27,#T_cac7f_row0_col28,#T_cac7f_row0_col29{\n",
       "            background-color:  #228a44;\n",
       "        }#T_cac7f_row0_col30,#T_cac7f_row0_col31,#T_cac7f_row0_col32{\n",
       "            background-color:  #712c6d;\n",
       "        }</style><table id=\"T_cac7f_\" ><thead>    <tr>        <th class=\"index_name level0\" >sample</th>        <th class=\"col_heading level0 col0\" >14</th>        <th class=\"col_heading level0 col1\" >15</th>        <th class=\"col_heading level0 col2\" >16</th>        <th class=\"col_heading level0 col3\" >29</th>        <th class=\"col_heading level0 col4\" >30</th>        <th class=\"col_heading level0 col5\" >31</th>        <th class=\"col_heading level0 col6\" >8</th>        <th class=\"col_heading level0 col7\" >9</th>        <th class=\"col_heading level0 col8\" >10</th>        <th class=\"col_heading level0 col9\" >23</th>        <th class=\"col_heading level0 col10\" >24</th>        <th class=\"col_heading level0 col11\" >25</th>        <th class=\"col_heading level0 col12\" >38</th>        <th class=\"col_heading level0 col13\" >39</th>        <th class=\"col_heading level0 col14\" >40</th>        <th class=\"col_heading level0 col15\" >17</th>        <th class=\"col_heading level0 col16\" >18</th>        <th class=\"col_heading level0 col17\" >19</th>        <th class=\"col_heading level0 col18\" >26</th>        <th class=\"col_heading level0 col19\" >27</th>        <th class=\"col_heading level0 col20\" >28</th>        <th class=\"col_heading level0 col21\" >20</th>        <th class=\"col_heading level0 col22\" >21</th>        <th class=\"col_heading level0 col23\" >22</th>        <th class=\"col_heading level0 col24\" >35</th>        <th class=\"col_heading level0 col25\" >36</th>        <th class=\"col_heading level0 col26\" >37</th>        <th class=\"col_heading level0 col27\" >11</th>        <th class=\"col_heading level0 col28\" >12</th>        <th class=\"col_heading level0 col29\" >13</th>        <th class=\"col_heading level0 col30\" >32</th>        <th class=\"col_heading level0 col31\" >33</th>        <th class=\"col_heading level0 col32\" >34</th>    </tr></thead><tbody>\n",
       "                <tr>\n",
       "                        <th id=\"T_cac7f_level0_row0\" class=\"row_heading level0 row0\" >label</th>\n",
       "                        <td id=\"T_cac7f_row0_col0\" class=\"data row0 col0\" >CAN</td>\n",
       "                        <td id=\"T_cac7f_row0_col1\" class=\"data row0 col1\" >CAN</td>\n",
       "                        <td id=\"T_cac7f_row0_col2\" class=\"data row0 col2\" >CAN</td>\n",
       "                        <td id=\"T_cac7f_row0_col3\" class=\"data row0 col3\" >CS</td>\n",
       "                        <td id=\"T_cac7f_row0_col4\" class=\"data row0 col4\" >CS</td>\n",
       "                        <td id=\"T_cac7f_row0_col5\" class=\"data row0 col5\" >CS</td>\n",
       "                        <td id=\"T_cac7f_row0_col6\" class=\"data row0 col6\" >LAB</td>\n",
       "                        <td id=\"T_cac7f_row0_col7\" class=\"data row0 col7\" >LAB</td>\n",
       "                        <td id=\"T_cac7f_row0_col8\" class=\"data row0 col8\" >LAB</td>\n",
       "                        <td id=\"T_cac7f_row0_col9\" class=\"data row0 col9\" >PN</td>\n",
       "                        <td id=\"T_cac7f_row0_col10\" class=\"data row0 col10\" >PN</td>\n",
       "                        <td id=\"T_cac7f_row0_col11\" class=\"data row0 col11\" >PN</td>\n",
       "                        <td id=\"T_cac7f_row0_col12\" class=\"data row0 col12\" >REG</td>\n",
       "                        <td id=\"T_cac7f_row0_col13\" class=\"data row0 col13\" >REG</td>\n",
       "                        <td id=\"T_cac7f_row0_col14\" class=\"data row0 col14\" >REG</td>\n",
       "                        <td id=\"T_cac7f_row0_col15\" class=\"data row0 col15\" >RIP</td>\n",
       "                        <td id=\"T_cac7f_row0_col16\" class=\"data row0 col16\" >RIP</td>\n",
       "                        <td id=\"T_cac7f_row0_col17\" class=\"data row0 col17\" >RIP</td>\n",
       "                        <td id=\"T_cac7f_row0_col18\" class=\"data row0 col18\" >RL</td>\n",
       "                        <td id=\"T_cac7f_row0_col19\" class=\"data row0 col19\" >RL</td>\n",
       "                        <td id=\"T_cac7f_row0_col20\" class=\"data row0 col20\" >RL</td>\n",
       "                        <td id=\"T_cac7f_row0_col21\" class=\"data row0 col21\" >ROT</td>\n",
       "                        <td id=\"T_cac7f_row0_col22\" class=\"data row0 col22\" >ROT</td>\n",
       "                        <td id=\"T_cac7f_row0_col23\" class=\"data row0 col23\" >ROT</td>\n",
       "                        <td id=\"T_cac7f_row0_col24\" class=\"data row0 col24\" >RU</td>\n",
       "                        <td id=\"T_cac7f_row0_col25\" class=\"data row0 col25\" >RU</td>\n",
       "                        <td id=\"T_cac7f_row0_col26\" class=\"data row0 col26\" >RU</td>\n",
       "                        <td id=\"T_cac7f_row0_col27\" class=\"data row0 col27\" >SYL</td>\n",
       "                        <td id=\"T_cac7f_row0_col28\" class=\"data row0 col28\" >SYL</td>\n",
       "                        <td id=\"T_cac7f_row0_col29\" class=\"data row0 col29\" >SYL</td>\n",
       "                        <td id=\"T_cac7f_row0_col30\" class=\"data row0 col30\" >TRI</td>\n",
       "                        <td id=\"T_cac7f_row0_col31\" class=\"data row0 col31\" >TRI</td>\n",
       "                        <td id=\"T_cac7f_row0_col32\" class=\"data row0 col32\" >TRI</td>\n",
       "            </tr>\n",
       "    </tbody></table>"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x1f4d5b1aa30>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parsed = mtl.parse_data(datasets['GD_pos_class2']['data'], labels_loc='label')\n",
    "y = datasets['GD_pos_class2']['target']\n",
    "label_colors = datasets['GD_pos_class2']['label_colors']\n",
    "s = styled_sample_labels(parsed.sample_names, y, label_colors)\n",
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "#T_1e6b4_row0_col0,#T_1e6b4_row0_col1,#T_1e6b4_row0_col2{\n",
       "            background-color:  #e41a1c;\n",
       "        }#T_1e6b4_row0_col3,#T_1e6b4_row0_col4,#T_1e6b4_row0_col5{\n",
       "            background-color:  #377eb8;\n",
       "        }#T_1e6b4_row0_col6,#T_1e6b4_row0_col7,#T_1e6b4_row0_col8{\n",
       "            background-color:  #4daf4a;\n",
       "        }#T_1e6b4_row0_col9,#T_1e6b4_row0_col10,#T_1e6b4_row0_col11{\n",
       "            background-color:  #984ea3;\n",
       "        }#T_1e6b4_row0_col12,#T_1e6b4_row0_col13,#T_1e6b4_row0_col14{\n",
       "            background-color:  #ff7f00;\n",
       "        }</style><table id=\"T_1e6b4_\" ><thead>    <tr>        <th class=\"index_name level0\" >sample</th>        <th class=\"col_heading level0 col0\" >BY0_1</th>        <th class=\"col_heading level0 col1\" >BY0_2</th>        <th class=\"col_heading level0 col2\" >BY0_3</th>        <th class=\"col_heading level0 col3\" >GRE3_1</th>        <th class=\"col_heading level0 col4\" >GRE3_2</th>        <th class=\"col_heading level0 col5\" >GRE3_3</th>        <th class=\"col_heading level0 col6\" >ENO1_1</th>        <th class=\"col_heading level0 col7\" >ENO1_2</th>        <th class=\"col_heading level0 col8\" >ENO1_3</th>        <th class=\"col_heading level0 col9\" >dGLO1_1</th>        <th class=\"col_heading level0 col10\" >dGLO1_2</th>        <th class=\"col_heading level0 col11\" >dGLO1_3</th>        <th class=\"col_heading level0 col12\" >GLO2_1</th>        <th class=\"col_heading level0 col13\" >GLO2_2</th>        <th class=\"col_heading level0 col14\" >GLO2_3</th>    </tr></thead><tbody>\n",
       "                <tr>\n",
       "                        <th id=\"T_1e6b4_level0_row0\" class=\"row_heading level0 row0\" >label</th>\n",
       "                        <td id=\"T_1e6b4_row0_col0\" class=\"data row0 col0\" >WT</td>\n",
       "                        <td id=\"T_1e6b4_row0_col1\" class=\"data row0 col1\" >WT</td>\n",
       "                        <td id=\"T_1e6b4_row0_col2\" class=\"data row0 col2\" >WT</td>\n",
       "                        <td id=\"T_1e6b4_row0_col3\" class=\"data row0 col3\" >ΔGRE3</td>\n",
       "                        <td id=\"T_1e6b4_row0_col4\" class=\"data row0 col4\" >ΔGRE3</td>\n",
       "                        <td id=\"T_1e6b4_row0_col5\" class=\"data row0 col5\" >ΔGRE3</td>\n",
       "                        <td id=\"T_1e6b4_row0_col6\" class=\"data row0 col6\" >ΔENO1</td>\n",
       "                        <td id=\"T_1e6b4_row0_col7\" class=\"data row0 col7\" >ΔENO1</td>\n",
       "                        <td id=\"T_1e6b4_row0_col8\" class=\"data row0 col8\" >ΔENO1</td>\n",
       "                        <td id=\"T_1e6b4_row0_col9\" class=\"data row0 col9\" >ΔGLO1</td>\n",
       "                        <td id=\"T_1e6b4_row0_col10\" class=\"data row0 col10\" >ΔGLO1</td>\n",
       "                        <td id=\"T_1e6b4_row0_col11\" class=\"data row0 col11\" >ΔGLO1</td>\n",
       "                        <td id=\"T_1e6b4_row0_col12\" class=\"data row0 col12\" >ΔGLO2</td>\n",
       "                        <td id=\"T_1e6b4_row0_col13\" class=\"data row0 col13\" >ΔGLO2</td>\n",
       "                        <td id=\"T_1e6b4_row0_col14\" class=\"data row0 col14\" >ΔGLO2</td>\n",
       "            </tr>\n",
       "    </tbody></table>"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x1f4d5b77190>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parsed = mtl.parse_data(datasets['YD']['data'])\n",
    "y = datasets['YD']['target']\n",
    "label_colors = datasets['YD']['label_colors']\n",
    "s = styled_sample_labels(parsed.sample_names, y, label_colors)\n",
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "#T_946e7_row0_col0,#T_946e7_row0_col1,#T_946e7_row0_col2,#T_946e7_row0_col6,#T_946e7_row0_col7,#T_946e7_row0_col8,#T_946e7_row0_col15,#T_946e7_row0_col16,#T_946e7_row0_col17,#T_946e7_row0_col21,#T_946e7_row0_col22,#T_946e7_row0_col23,#T_946e7_row0_col24,#T_946e7_row0_col25,#T_946e7_row0_col26,#T_946e7_row0_col27,#T_946e7_row0_col28,#T_946e7_row0_col29{\n",
       "            background-color:  #228a44;\n",
       "        }#T_946e7_row0_col3,#T_946e7_row0_col4,#T_946e7_row0_col5,#T_946e7_row0_col9,#T_946e7_row0_col10,#T_946e7_row0_col11,#T_946e7_row0_col12,#T_946e7_row0_col13,#T_946e7_row0_col14,#T_946e7_row0_col18,#T_946e7_row0_col19,#T_946e7_row0_col20,#T_946e7_row0_col30,#T_946e7_row0_col31,#T_946e7_row0_col32{\n",
       "            background-color:  #712c6d;\n",
       "        }</style><table id=\"T_946e7_\" ><thead>    <tr>        <th class=\"index_name level0\" >sample</th>        <th class=\"col_heading level0 col0\" >14</th>        <th class=\"col_heading level0 col1\" >15</th>        <th class=\"col_heading level0 col2\" >16</th>        <th class=\"col_heading level0 col3\" >29</th>        <th class=\"col_heading level0 col4\" >30</th>        <th class=\"col_heading level0 col5\" >31</th>        <th class=\"col_heading level0 col6\" >8</th>        <th class=\"col_heading level0 col7\" >9</th>        <th class=\"col_heading level0 col8\" >10</th>        <th class=\"col_heading level0 col9\" >23</th>        <th class=\"col_heading level0 col10\" >24</th>        <th class=\"col_heading level0 col11\" >25</th>        <th class=\"col_heading level0 col12\" >38</th>        <th class=\"col_heading level0 col13\" >39</th>        <th class=\"col_heading level0 col14\" >40</th>        <th class=\"col_heading level0 col15\" >17</th>        <th class=\"col_heading level0 col16\" >18</th>        <th class=\"col_heading level0 col17\" >19</th>        <th class=\"col_heading level0 col18\" >26</th>        <th class=\"col_heading level0 col19\" >27</th>        <th class=\"col_heading level0 col20\" >28</th>        <th class=\"col_heading level0 col21\" >20</th>        <th class=\"col_heading level0 col22\" >21</th>        <th class=\"col_heading level0 col23\" >22</th>        <th class=\"col_heading level0 col24\" >35</th>        <th class=\"col_heading level0 col25\" >36</th>        <th class=\"col_heading level0 col26\" >37</th>        <th class=\"col_heading level0 col27\" >11</th>        <th class=\"col_heading level0 col28\" >12</th>        <th class=\"col_heading level0 col29\" >13</th>        <th class=\"col_heading level0 col30\" >32</th>        <th class=\"col_heading level0 col31\" >33</th>        <th class=\"col_heading level0 col32\" >34</th>    </tr></thead><tbody>\n",
       "                <tr>\n",
       "                        <th id=\"T_946e7_level0_row0\" class=\"row_heading level0 row0\" >label</th>\n",
       "                        <td id=\"T_946e7_row0_col0\" class=\"data row0 col0\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col1\" class=\"data row0 col1\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col2\" class=\"data row0 col2\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col3\" class=\"data row0 col3\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col4\" class=\"data row0 col4\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col5\" class=\"data row0 col5\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col6\" class=\"data row0 col6\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col7\" class=\"data row0 col7\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col8\" class=\"data row0 col8\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col9\" class=\"data row0 col9\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col10\" class=\"data row0 col10\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col11\" class=\"data row0 col11\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col12\" class=\"data row0 col12\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col13\" class=\"data row0 col13\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col14\" class=\"data row0 col14\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col15\" class=\"data row0 col15\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col16\" class=\"data row0 col16\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col17\" class=\"data row0 col17\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col18\" class=\"data row0 col18\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col19\" class=\"data row0 col19\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col20\" class=\"data row0 col20\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col21\" class=\"data row0 col21\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col22\" class=\"data row0 col22\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col23\" class=\"data row0 col23\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col24\" class=\"data row0 col24\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col25\" class=\"data row0 col25\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col26\" class=\"data row0 col26\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col27\" class=\"data row0 col27\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col28\" class=\"data row0 col28\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col29\" class=\"data row0 col29\" >wild</td>\n",
       "                        <td id=\"T_946e7_row0_col30\" class=\"data row0 col30\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col31\" class=\"data row0 col31\" >vinifera</td>\n",
       "                        <td id=\"T_946e7_row0_col32\" class=\"data row0 col32\" >vinifera</td>\n",
       "            </tr>\n",
       "    </tbody></table>"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x1f4d5e55f70>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parsed = mtl.parse_data(datasets['vitis_types']['data'], labels_loc='label')\n",
    "y = datasets['vitis_types']['target']\n",
    "label_colors = datasets['vitis_types']['label_colors']\n",
    "s = styled_sample_labels(parsed.sample_names, y, label_colors)\n",
    "s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data transformations and pre-treatments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Each Dataset is transformed to binary or treated by a combination of more established treatments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Traditional Pre-Treatments\n",
    "\n",
    "All samples need to have missing value imputation (removing features is optional). In this case, missing features were replaced with half of the minimum intensity of the full dataset.\n",
    "\n",
    "This can be done with the `NaN_Imputation` function of scaling.py or `fillna_frac_min`(more options available) from metabolinks.\n",
    "\n",
    "#### 3/4 types of combinations of traditional pre-treatments are made:\n",
    "\n",
    "(Imp (Ionly) - Only Missing Value Imputation).\n",
    "\n",
    "##### P Treatment - Missing Value Imputation and Pareto Scaling.\n",
    "\n",
    "##### NP Treatment - Missing Value Imputation, Normalization by reference feature and Pareto Scaling.\n",
    "\n",
    "##### NGP Treatment - Missing Value Imputation, Normalization by reference feature, Generalized Logarithmic Transformation and Pareto Scaling.\n",
    "\n",
    "Note: Leucine Enkephalin peak is removed upon normalization.\n",
    "\n",
    "### Binary Simplification (BinSim)\n",
    "\n",
    "##### BinSim - `df_to_bool` function (very simple)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Represents Binary Similarity pre-treatment\n",
    "def df_to_bool(df):\n",
    "    \"Transforms data into 'binary' matrices.\"\n",
    "    return df.mask(df.notnull(), 1).mask(df.isnull(), 0)\n",
    "\n",
    "\n",
    "def compute_transf(dataset, norm_ref=None, lamb=None):\n",
    "    \"Computes 3 combinations of pre-treatments and BinSim and returns after treatment datasets in a dict.\"\n",
    "    \n",
    "    data = dataset['data']\n",
    "    \n",
    "    # Imputation of Missing Values\n",
    "    imputed = transf.fillna_frac_min(data, fraction=0.5)\n",
    "    \n",
    "    # Normalization by a reference feature\n",
    "    if norm_ref is not None:\n",
    "        norm = transf.normalize_ref_feature(imputed, norm_ref, remove=True)\n",
    "    else:\n",
    "        norm = imputed\n",
    "    \n",
    "    # Pareto Scaling and Generalized Logarithmic Transformation\n",
    "    P = transf.pareto_scale(imputed)\n",
    "    NP = transf.pareto_scale(norm)\n",
    "    NGP = transf.pareto_scale(transf.glog(norm, lamb=lamb))\n",
    "    \n",
    "    # Store results\n",
    "    dataset['BinSim'] = df_to_bool(data)\n",
    "    dataset['Ionly'] = imputed\n",
    "    dataset['P'] = P\n",
    "    dataset['NP'] = NP\n",
    "    dataset['NGP'] = NGP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Human leucine enkephalin (Sigma Aldrich) is the reference feature (internal standard) used for these datasets\n",
    "\n",
    "[M+H]+ = 556.276575 Da or [M-H]- = 554.262022 Da\n",
    "\n",
    "Search in the Negative GD and Positive GD for the reference feature with `search_for_ref_feat` (details in scaling.py).\n",
    "\n",
    "### 1st Filtering/Alignment - 1-2 (used in dissertation)\n",
    "\n",
    "#### Performing different combinations of pre-treatments for the Negative and Positive GD\n",
    "\n",
    "Search for the reference feature."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Theoretical mass for negative mode Leucine Enkephalin - 554.262022\n",
    "# Theoretical mass for positive mode Leucine Enkephalin - 556.276575\n",
    "Leu_Enk_neg = 554.262022\n",
    "Leu_Enk_pos = 556.276575\n",
    "\n",
    "# Reference Feature in the yeast dataset\n",
    "leu_enk_name = '555.2692975341 Da'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "looking for reference in GD_neg_global2 ...\n",
      "Found ref feature 554.262022\n",
      "In data: 554.26202 delta = -2.000e-06\n",
      "\n",
      "looking for reference in GD_pos_global2 ...\n",
      "Found ref feature 556.276575\n",
      "In data: 556.27657 delta = -5.000e-06\n",
      "\n",
      "looking for reference in GD_neg_class2 ...\n",
      "Found ref feature 554.262022\n",
      "In data: 554.26202 delta = -2.000e-06\n",
      "\n",
      "looking for reference in GD_pos_class2 ...\n",
      "Found ref feature 556.276575\n",
      "In data: 556.27657 delta = -5.000e-06\n",
      "\n",
      "looking for reference in vitis_types ...\n",
      "Found ref feature 554.262022\n",
      "In data: 554.26202 delta = -2.000e-06\n",
      "\n",
      "looking for reference in YD ...\n",
      "Found ref feature 555.2692975341 Da \n",
      "\n",
      "looking for reference in YD2 ...\n",
      "Found ref feature 555.2692975341 Da \n",
      "\n"
     ]
    }
   ],
   "source": [
    "query_datasets = [name for name in datasets if datasets[name]['source']=='grapevine']\n",
    "\n",
    "for name in query_datasets:\n",
    "    ds = datasets[name]\n",
    "    print(f'looking for reference in {name} ...')\n",
    "    ref_variable = Leu_Enk_neg if ds['mode'] == '-' else Leu_Enk_pos\n",
    "    closest = transf.find_closest_features(ds['data'], features=[ref_variable])\n",
    "    if closest[ref_variable] is not None:\n",
    "        print('Found ref feature', ref_variable)\n",
    "        delta = closest[ref_variable] - ref_variable\n",
    "        print(f'In data: {closest[ref_variable]} delta = {delta:.3e}\\n')\n",
    "\n",
    "query_datasets = [name for name in datasets if name.startswith('YD')]\n",
    "\n",
    "ref_variable = leu_enk_name\n",
    "for name in query_datasets:\n",
    "    ds = datasets[name]\n",
    "    print(f'looking for reference in {name} ...') \n",
    "    closest = transf.find_closest_features(ds['data'], features=[ref_variable])\n",
    "    if closest[ref_variable] is not None:\n",
    "        print('Found ref feature', ref_variable, '\\n')\n",
    "    else:\n",
    "        print('Ref feature not found\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Apply the different pre-treatments and get the results in their respective dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying pre-processing transformations to data in GD_neg_global2 ...done!\n",
      "Applying pre-processing transformations to data in GD_pos_global2 ...done!\n",
      "Applying pre-processing transformations to data in GD_neg_class2 ...done!\n",
      "Applying pre-processing transformations to data in GD_pos_class2 ...done!\n",
      "Applying pre-processing transformations to data in YD ...done!\n",
      "Applying pre-processing transformations to data in YD2 ...done!\n",
      "Applying pre-processing transformations to data in vitis_types ...done!\n"
     ]
    }
   ],
   "source": [
    "for name, ds in datasets.items():\n",
    "    print(f'Applying pre-processing transformations to data in {name}', end=' ...')\n",
    "    if name.startswith('YD'):\n",
    "        ref_variable = leu_enk_name\n",
    "    else:\n",
    "        ref_variable = Leu_Enk_neg if ds['mode'] == '-' else Leu_Enk_pos\n",
    "    \n",
    "    compute_transf(ds, norm_ref=ref_variable)\n",
    "    print('done!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supervised Statistical Analysis\n",
    "\n",
    "The Supervised Statistical Analysis methods used will be Random Forest and PLS-DA.\n",
    "\n",
    "The performance of the classifiers will be evaluated by their predictive **accuracy** (which will always be estimated by internal stratified 3-fold cross-validation or 5 -fold cross-validation in vitis_types)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Permutation Tests - Part of Suppl. Fig. 6.7\n",
    "\n",
    "Permutation tests is based on shuffling the labels of the different samples, shuflling the groups where they belong with the intent to see if the classifier tested, whether it is Random Forest or PLS-DA found a significant class structure in the data - assess the significance of the predictive accuracy results. \n",
    "\n",
    "For that a random 3-fold cross-validation is performed on the original dataset (to serve as a comparation point) and 1000 permutations of datasets with labels randomly shuffled around with the model being evaluated by how good the predictions the model does of the test data is. \n",
    "\n",
    "The empirical p-value is given by (the number of times the permutation accuracy was bigger than the random 3-fold cross-validation made with the original dataset + 1) / (number of permutations + 1) (source: Ojala2010 - error should be the opposite of the accuracy).\n",
    "\n",
    "Histograms with the prediction accuracy of the different permutations were plotted and compared to the accuracy got with the original dataset.\n",
    "\n",
    "### Permutation Tests - Random Forests (very slow)\n",
    "\n",
    "Use of `permutation_RF` function from multianalysis.py. See details about the application of this function in the multianalysis.py file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# json for persistence\n",
    "\n",
    "import json\n",
    "from time import perf_counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500 permutations (Random Forest) for GD_neg_global2 with treatment P ...Done! took 203.279 s\n",
      "500 permutations (Random Forest) for GD_neg_global2 with treatment NP ...Done! took 217.011 s\n",
      "500 permutations (Random Forest) for GD_neg_global2 with treatment NGP ...Done! took 217.454 s\n",
      "500 permutations (Random Forest) for GD_neg_global2 with treatment BinSim ...Done! took 205.516 s\n",
      "500 permutations (Random Forest) for GD_pos_global2 with treatment P ...Done! took 249.120 s\n",
      "500 permutations (Random Forest) for GD_pos_global2 with treatment NP ...Done! took 279.882 s\n",
      "500 permutations (Random Forest) for GD_pos_global2 with treatment NGP ...Done! took 274.214 s\n",
      "500 permutations (Random Forest) for GD_pos_global2 with treatment BinSim ...Done! took 248.160 s\n",
      "500 permutations (Random Forest) for GD_neg_class2 with treatment P ...Done! took 192.610 s\n",
      "500 permutations (Random Forest) for GD_neg_class2 with treatment NP ...Done! took 203.983 s\n",
      "500 permutations (Random Forest) for GD_neg_class2 with treatment NGP ...Done! took 207.553 s\n",
      "500 permutations (Random Forest) for GD_neg_class2 with treatment BinSim ...Done! took 191.032 s\n",
      "500 permutations (Random Forest) for GD_pos_class2 with treatment P ...Done! took 219.905 s\n",
      "500 permutations (Random Forest) for GD_pos_class2 with treatment NP ...Done! took 239.085 s\n",
      "500 permutations (Random Forest) for GD_pos_class2 with treatment NGP ...Done! took 229.138 s\n",
      "500 permutations (Random Forest) for GD_pos_class2 with treatment BinSim ...Done! took 208.355 s\n",
      "500 permutations (Random Forest) for YD with treatment P ...Done! took 167.753 s\n",
      "500 permutations (Random Forest) for YD with treatment NP ...Done! took 169.997 s\n",
      "500 permutations (Random Forest) for YD with treatment NGP ...Done! took 168.293 s\n",
      "500 permutations (Random Forest) for YD with treatment BinSim ...Done! took 165.744 s\n",
      "500 permutations (Random Forest) for YD2 with treatment P ...Done! took 149.495 s\n",
      "500 permutations (Random Forest) for YD2 with treatment NP ...Done! took 149.414 s\n",
      "500 permutations (Random Forest) for YD2 with treatment NGP ...Done! took 150.688 s\n",
      "500 permutations (Random Forest) for YD2 with treatment BinSim ...Done! took 148.052 s\n",
      "500 permutations (Random Forest) for vitis_types with treatment P ...Done! took 309.952 s\n",
      "500 permutations (Random Forest) for vitis_types with treatment NP ...Done! took 1606.163 s\n",
      "500 permutations (Random Forest) for vitis_types with treatment NGP ...Done! took 335.034 s\n",
      "500 permutations (Random Forest) for vitis_types with treatment BinSim ...Done! took 337.354 s\n"
     ]
    }
   ],
   "source": [
    "# NOTE: for debugging\n",
    "iter_num=500\n",
    "# otherwise\n",
    "#iter_num=1000\n",
    "\n",
    "\n",
    "permuts_RF = []\n",
    "\n",
    "to_permute = [name for name in datasets]# if 'global2' in name]\n",
    "for name in to_permute:\n",
    "    for treatment in ('P', 'NP', 'NGP', 'BinSim'):\n",
    "        dataset = datasets[name]\n",
    "        print(f'{iter_num} permutations (Random Forest) for {name} with treatment {treatment}', end=' ...')\n",
    "        n_fold = 5 if name == 'vitis_types' else 3\n",
    "        start = perf_counter()\n",
    "        permutations = ma.permutation_RF(dataset[treatment], dataset['target'], iter_num=iter_num, n_fold=n_fold, n_trees=100)\n",
    "        res = {'dataset': name, 'treatment': treatment,\n",
    "               'non_permuted_CV': permutations[0],\n",
    "               'permutations': permutations[1],\n",
    "               'p-value': permutations[2]}       \n",
    "        permuts_RF.append(res)\n",
    "        end = perf_counter()\n",
    "        print(f'Done! took {(end - start):.3f} s')\n",
    "\n",
    "fname = 'permuts/permuts_rf.json'\n",
    "with open(fname, \"w\", encoding='utf8') as write_file:\n",
    "    json.dump(permuts_RF, write_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Permutation Tests - PLS-DA (slow)\n",
    "\n",
    "Use of `permutation_PLSDA` function from multianalysis.py. See details about the application of this function in the multianalysis.py file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Permutation test (PLS-DA) for GD_neg_global2 with treatment P ...Done! took 141.125 s\n",
      "Permutation test (PLS-DA) for GD_neg_global2 with treatment NP ...Done! took 135.278 s\n",
      "Permutation test (PLS-DA) for GD_neg_global2 with treatment NGP ...Done! took 129.434 s\n",
      "Permutation test (PLS-DA) for GD_neg_global2 with treatment BinSim ...Done! took 141.853 s\n",
      "Permutation test (PLS-DA) for GD_pos_global2 with treatment P ...Done! took 222.169 s\n",
      "Permutation test (PLS-DA) for GD_pos_global2 with treatment NP ...Done! took 233.709 s\n",
      "Permutation test (PLS-DA) for GD_pos_global2 with treatment NGP ...Done! took 237.552 s\n",
      "Permutation test (PLS-DA) for GD_pos_global2 with treatment BinSim ...Done! took 255.252 s\n",
      "Permutation test (PLS-DA) for GD_neg_class2 with treatment P ...Done! took 105.750 s\n",
      "Permutation test (PLS-DA) for GD_neg_class2 with treatment NP ...Done! took 112.427 s\n",
      "Permutation test (PLS-DA) for GD_neg_class2 with treatment NGP ...Done! took 123.929 s\n",
      "Permutation test (PLS-DA) for GD_neg_class2 with treatment BinSim ...Done! took 128.859 s\n",
      "Permutation test (PLS-DA) for GD_pos_class2 with treatment P ...Done! took 163.595 s\n",
      "Permutation test (PLS-DA) for GD_pos_class2 with treatment NP ...Done! took 164.911 s\n",
      "Permutation test (PLS-DA) for GD_pos_class2 with treatment NGP ...Done! took 165.765 s\n",
      "Permutation test (PLS-DA) for GD_pos_class2 with treatment BinSim ...Done! took 156.941 s\n",
      "Permutation test (PLS-DA) for YD with treatment P ...Done! took 54.788 s\n",
      "Permutation test (PLS-DA) for YD with treatment NP ...Done! took 53.866 s\n",
      "Permutation test (PLS-DA) for YD with treatment NGP ...Done! took 53.488 s\n",
      "Permutation test (PLS-DA) for YD with treatment BinSim ...Done! took 57.840 s\n",
      "Permutation test (PLS-DA) for YD2 with treatment P ...Done! took 27.884 s\n",
      "Permutation test (PLS-DA) for YD2 with treatment NP ...Done! took 27.118 s\n",
      "Permutation test (PLS-DA) for YD2 with treatment NGP ...Done! took 27.263 s\n",
      "Permutation test (PLS-DA) for YD2 with treatment BinSim ...Done! took 30.196 s\n",
      "Permutation test (PLS-DA) for vitis_types with treatment P ...Done! took 114.889 s\n",
      "Permutation test (PLS-DA) for vitis_types with treatment NP ...Done! took 114.803 s\n",
      "Permutation test (PLS-DA) for vitis_types with treatment NGP ...Done! took 114.115 s\n",
      "Permutation test (PLS-DA) for vitis_types with treatment BinSim ...Done! took 112.117 s\n"
     ]
    }
   ],
   "source": [
    "%%capture --no-stdout\n",
    "\n",
    "# NOTE: for debugging\n",
    "iter_num=500\n",
    "# otherwise\n",
    "#iter_num=1000\n",
    "\n",
    "\n",
    "permuts_PLSDA = []\n",
    "\n",
    "to_permute = [name for name in datasets] # if 'global2' in name]\n",
    "for name in to_permute:\n",
    "    for treatment in ('P', 'NP', 'NGP', 'BinSim'):\n",
    "        dataset = datasets[name]\n",
    "        print(f'Permutation test (PLS-DA) for {name} with treatment {treatment}', end=' ...')\n",
    "        n_comp = 11 if name.startswith('GD') else 6\n",
    "        n_fold = 5 if name == 'vitis_types' else 3\n",
    "        start = perf_counter()\n",
    "        permutations = ma.permutation_PLSDA(dataset[treatment], dataset['target'], n_comp=11,\n",
    "                                            iter_num=iter_num, n_fold=n_fold)\n",
    "        res = {'dataset': name, 'treatment': treatment,\n",
    "               'non_permuted_CV': permutations[0],\n",
    "               'permutations': permutations[1],\n",
    "               'p-value': permutations[2]}\n",
    "        permuts_PLSDA.append(res)\n",
    "        end = perf_counter()\n",
    "        print(f'Done! took {(end - start):.3f} s')\n",
    "fname = 'permuts/permuts_plsda.json'\n",
    "with open(fname, \"w\", encoding='utf8') as write_file:\n",
    "    json.dump(permuts_PLSDA, write_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
